{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#!/usr/bin/env py27\n",
    "import warnings\n",
    "with warnings.catch_warnings():\n",
    "    warnings.simplefilter('ignore')\n",
    "from IPython.display import display, Image\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "#import pythoncyc as pc\n",
    "#from HTMLParser import HTMLParser\n",
    "#from lxml import html\n",
    "import Bio\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import generic_dna\n",
    "import csv\n",
    "import pandas as pd\n",
    "import math\n",
    "import re\n",
    "import numpy as np\n",
    "\n",
    "# this function finds all pathways associated with the molecule of interest. needs work on front end to \n",
    "# allow user to define G.O. function associated with molecule.\n",
    "class organismBreadBoard:\n",
    "    def __init__(self, name):\n",
    "        self.name = name\n",
    "    def pathwayFromCompound(self, compound):\n",
    "        self.compound = compound\n",
    "        pathWays = {}\n",
    "    #request pathway xml from ecocyc server\n",
    "        resp = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-compound&id=ECOLI:{}\".format(self.compound))\n",
    "    #builds element tree from the response xml text file\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #pulls all elements associated with Pathway tag\n",
    "        pathways = root.findall(\"./Pathway\")\n",
    "    #pulls all pathway IDs\n",
    "        for pathway in pathways:\n",
    "            pathID = pathway.attrib['frameid']\n",
    "            pathWays[pathID] = pathway.findall(\"common-name\")[0].text\n",
    "        #print (pathWays)\n",
    "    #return dictionary with pathway ID: common name as key:value pairs\n",
    "    #print pathWays\n",
    "        return pathWays\n",
    "\n",
    "# this function pulls all geneIDs associated with the pathwayID passed into it\n",
    "    def genesFromPathway(self, pathwayID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=genes-of-pathway&id=ECOLI:{}&detail=full\".format(pathwayID))\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        genes = root.findall(\"./Gene\")\n",
    "        genesDict = {}\n",
    "        for gene in genes:\n",
    "            geneID = gene.attrib['frameid']\n",
    "            geneName = gene.findall(\"common-name\")[0].text\n",
    "            geneStrand= gene.findall(\"transcription-direction\")[0].text\n",
    "            #print geneStrand\n",
    "        #print \"Gene: \" + geneName + \", GeneID: \" + geneID\n",
    "        #page = requests.get(\"http://biocyc.org/gene?orgid=ECOLI&id=EG11879\")\n",
    "            genesDict[geneID] = [geneName, geneStrand]\n",
    "        return genesDict\n",
    "    \n",
    "# this function pulls the promoterIDs associated with the geneID passed into it\n",
    "    def getPromoter(self, geneID, strand):\n",
    "        promoterDict = {}\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        #print resp.text\n",
    "        #transUnits=root.findall(\"./Transcription-Unit\")\n",
    "        \n",
    "        promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "        #for transUnit in transUnits:\n",
    "            #TUID=transUnit.attrib['frameid']\n",
    "            #print TUID\n",
    "            #print promoters\n",
    "        for promoter in promoters:\n",
    "            #print promoters[promoter]\n",
    "            pID = promoter.attrib['frameid']\n",
    "            print pID\n",
    "            TUID = promoter.find('transcription-unit')\n",
    "            print TUID\n",
    "                #promoterDict[pID]=[TUID]\n",
    "\n",
    "            if pID not in promoterDict:\n",
    "                try:\n",
    "                    #absPositPlusOne = promoter.find('absolute-plus-1-pos').text\n",
    "                    #promoterDict[pID]= absPositPlusOne\n",
    "                    promoterDict[pID]=[promoter.find('absolute-plus-1-pos').text, strand]\n",
    "                    #print 'promoter position for:', pID,  'is:', promoterDict[pID]\n",
    "                #print('b')\n",
    "                except AttributeError:\n",
    "                    promoterDict[pID]=['NA','NA']\n",
    "                    #print 'promoter position for',pID, 'was not found'\n",
    "            else:\n",
    "                \n",
    "                continue\n",
    "            #promoterDict[pID]= absPositPlusOne\n",
    "        print promoterDict\n",
    "        return promoterDict\n",
    "    \n",
    "    def getTUID(self, geneID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        transUnits=root.findall(\"./Transcription-Unit\")\n",
    "        TU_list=[]\n",
    "        for transUnit in transUnits:\n",
    "            TU_list.append(transUnit.attrib['frameid'])\n",
    "        return TU_list\n",
    "    \n",
    "    def getPromoOfTU(self, TUID):\n",
    "        promoterDict={}\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-unit-promoter&id=ECOLI:{}&detail=full\".format(TUID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        #print resp.text\n",
    "        promoters=root.findall(\"./Promoter\")\n",
    "        #print promoters\n",
    "        for promoter in promoters:\n",
    "            pID = promoter.attrib['frameid']\n",
    "            trnsDir = promoter.findall('transcription-direction')[0].text\n",
    "            position = promoter.findall('absolute-plus-1-pos')[0].text\n",
    "            #print pID, trnsDir, position\n",
    "            promoterDict[pID]=[trnsDir, position]\n",
    "        return promoterDict\n",
    "    \n",
    "    def genesOfTU(self, TU_ID):\n",
    "        genesList = []\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-unit-genes&id=ECOLI:{}&detail=full\".format(TU_ID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        print(resp.text)\n",
    "        #genes = root.findall(\"./\")\n",
    "        \n",
    "    def pathwaysOfGene(self, geneID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        print(resp.text)\n",
    "        \n",
    "    def findAltFunc(self, pathwayID):\n",
    "        resp  = requests.get(\"http://metacyc.org/apixml?fn=pathways-of-gene&id=META:{}&detail=full\".format(pathwayID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        print(resp.text)\n",
    "        \n",
    "    def getGeneSeq(self, geneID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=get-gene-sequence&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        print(resp.text)\n",
    "        \n",
    "    def getGenePromoter(self, geneID, strand):\n",
    "        promoterDict = {}\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "        #print resp.text\n",
    "        promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "        for promoter in promoters:\n",
    "            pID = promoter.attrib['frameid']\n",
    "            if pID not in promoterDict:\n",
    "                try:\n",
    "                    promoterDict[pID]=[promoter.find('absolute-plus-1-pos').text, strand]\n",
    "                except AttributeError:\n",
    "                    promoterDict[pID]=['NA','NA']\n",
    "            else:\n",
    "                continue\n",
    "        return promoterDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "class FastAreader :\n",
    "    '''\n",
    "    Class to provide reading of a file containing one or more FASTA\n",
    "    formatted sequences:\n",
    "    object instantiation:\n",
    "    FastAreader(<file name>):\n",
    " \n",
    "    object attributes:\n",
    "    fname: the initial file name\n",
    " \n",
    "    methods:\n",
    "    readFasta() : returns header and sequence as strings.\n",
    "    Author: David Bernick\n",
    "    Date: April 19, 2013\n",
    "    '''\n",
    "    def __init__ (self, fname):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    " \n",
    "    def readFasta (self):\n",
    "        '''\n",
    "        using filename given in init, returns each included FastA record\n",
    "        as 2 strings - header and sequence.\n",
    "        whitespace is removed, no adjustment is made to sequence contents.\n",
    "        The initial '>' is removed from the header.\n",
    "        '''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with open(self.fname) as fileH:\n",
    "            # initialize return containers\n",
    "            header = ''\n",
    "            sequence = ''\n",
    " \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    " \n",
    "            # header is saved, get the rest of the sequence\n",
    "            # up until the next header is found\n",
    "            # then yield the results and wait for the next call.\n",
    "            # next call will resume at the yield point\n",
    "            # which is where we have the next header\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        # final header and sequence will be seen with an end of file\n",
    "        # with clause will terminate, so we do the final yield of the data\n",
    "        yield header,sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     Gene Name Accession-1  Left-End-Position  Right-End-Position  \\\n",
      "0        G7935       b4339          4569358.0           4569918.0   \n",
      "1        G7934       b4338          4568998.0           4569309.0   \n",
      "2        G7903       b4282          4506861.0           4507109.0   \n",
      "3        G7822       b4103          4322443.0           4322811.0   \n",
      "4        G7798       b3777          3960242.0           3960460.0   \n",
      "5        G7760       b3443          3583041.0           3583454.0   \n",
      "6        G7536       b2969          3111128.0           3111988.0   \n",
      "7        G7478       b2858          2995748.0           2996020.0   \n",
      "8        G7474       b2854          2994041.0           2994415.0   \n",
      "9        G7383       b2648          2783065.0           2783208.0   \n",
      "10       G7222       b2353          2470803.0           2471105.0   \n",
      "11       G7120       b2083          2167602.0           2167820.0   \n",
      "12       G7044       b1936          2012502.0           2012780.0   \n",
      "13       G6632       b1240          1294968.0           1295198.0   \n",
      "14       G6535       b1028          1095544.0           1095846.0   \n",
      "15       G6376       b0703           734220.0            735653.0   \n",
      "16       G6348       b0626           658255.0            658818.0   \n",
      "17       G6119       b0235           253467.0            253733.0   \n",
      "18       G6114       b0230           250072.0            250827.0   \n",
      "19       G6113       b0229           248358.0            250070.0   \n",
      "20    G0-10682       b4631           317439.0            317567.0   \n",
      "21    G0-10615       b4628           266776.0            266967.0   \n",
      "22    G0-10614       b4627           266553.0            266774.0   \n",
      "23    G0-10604       b4614          3722076.0           3722105.0   \n",
      "24    G0-10567       b4561          4508574.0           4508942.0   \n",
      "25    G0-10564       b4556          3861173.0           3861349.0   \n",
      "26    G0-10563       b4552          3623887.0           3624132.0   \n",
      "27    G0-10561       b4543          2357816.0           2358001.0   \n",
      "28    G0-10559       b4534          1652728.0           1652838.0   \n",
      "29    G0-10556       b4519          1211425.0           1211577.0   \n",
      "...        ...         ...                ...                 ...   \n",
      "4600   EG11175       b3124          3270625.0           3271770.0   \n",
      "4601   EG11101       b0476           499014.0            499973.0   \n",
      "4602   EG11075       b3564          3727917.0           3729371.0   \n",
      "4603   EG11074       b3565          3729443.0           3730765.0   \n",
      "4604   EG11066       b4322          4551636.0           4552820.0   \n",
      "4605   EG11065       b1521          1609229.0           1610680.0   \n",
      "4606   EG11055       b1617          1694260.0           1696071.0   \n",
      "4607   EG10971       b2705          2827737.0           2828516.0   \n",
      "4608   EG10818       b3752          3937294.0           3938223.0   \n",
      "4609   EG10817       b3748          3933351.0           3933770.0   \n",
      "4610   EG10700       b1723          1806370.0           1807299.0   \n",
      "4611   EG10699       b3916          4107552.0           4108514.0   \n",
      "4612   EG10616       b3600          3774424.0           3775572.0   \n",
      "4613   EG10577       b4119          4341911.0           4343266.0   \n",
      "4614   EG10559       b4037          4249554.0           4250474.0   \n",
      "4615   EG10527       b0344           363231.0            366305.0   \n",
      "4616   EG10524       b0342           361249.0            361860.0   \n",
      "4617   EG10381       b3431          3569346.0           3571319.0   \n",
      "4618   EG10355       b2804          2938888.0           2939310.0   \n",
      "4619   EG10350       b2803          2937438.0           2938886.0   \n",
      "4620   EG10349       b2802          2935584.0           2937359.0   \n",
      "4621   EG10348       b2800          2933041.0           2933688.0   \n",
      "4622   EG10282       b2925          3070165.0           3071244.0   \n",
      "4623   EG10144       b1734          1817148.0           1818500.0   \n",
      "4624   EG10114       b3721          3902289.0           3903701.0   \n",
      "4625   EG10085       b2716          2840990.0           2842414.0   \n",
      "4626   EG10055       b0061            65855.0             66550.0   \n",
      "4627   EG10053       b0063            68348.0             70048.0   \n",
      "4628   EG10052       b0062            66835.0             68337.0   \n",
      "4629   EG10016       b3126          3272787.0           3273557.0   \n",
      "\n",
      "                          Product  \\\n",
      "0                   G7935-MONOMER   \n",
      "1                   G7934-MONOMER   \n",
      "2                   G7903-MONOMER   \n",
      "3                   G7822-MONOMER   \n",
      "4                   G7798-MONOMER   \n",
      "5                   G7760-MONOMER   \n",
      "6                   G7536-MONOMER   \n",
      "7                   G7478-MONOMER   \n",
      "8                   G7474-MONOMER   \n",
      "9                   G7383-MONOMER   \n",
      "10                            NaN   \n",
      "11                  G7120-MONOMER   \n",
      "12                  G7044-MONOMER   \n",
      "13                  G6632-MONOMER   \n",
      "14                  G6535-MONOMER   \n",
      "15                  G6376-MONOMER   \n",
      "16                  G6348-MONOMER   \n",
      "17                  G6119-MONOMER   \n",
      "18                  G6114-MONOMER   \n",
      "19                  G6113-MONOMER   \n",
      "20                            NaN   \n",
      "21                            NaN   \n",
      "22                            NaN   \n",
      "23                            NaN   \n",
      "24                            NaN   \n",
      "25                            NaN   \n",
      "26                            NaN   \n",
      "27                            NaN   \n",
      "28                  MONOMER0-2781   \n",
      "29                  MONOMER0-2778   \n",
      "...                           ...   \n",
      "4600                  GKI-MONOMER   \n",
      "4601              EG11101-MONOMER   \n",
      "4602            XYLULOKIN-MONOMER   \n",
      "4603              XYLISOM-MONOMER   \n",
      "4604       MANNONDEHYDRAT-MONOMER   \n",
      "4605    ALTRO-OXIDOREDUCT-MONOMER   \n",
      "4606      BETA-GLUCURONID-MONOMER   \n",
      "4607       SORB6PDEHYDROG-MONOMER   \n",
      "4608              RIBOKIN-MONOMER   \n",
      "4609              EG10817-MONOMER   \n",
      "4610               6PFK-2-MONOMER   \n",
      "4611               6PFK-1-MONOMER   \n",
      "4612        MANNPDEHYDROG-MONOMER   \n",
      "4613      ALPHAGALACTOSID-MONOMER   \n",
      "4614              EG10559-MONOMER   \n",
      "4615       BETAGALACTOSID-MONOMER   \n",
      "4616    GALACTOACETYLTRAN-MONOMER   \n",
      "4617              EG10381-MONOMER   \n",
      "4618              EG10355-MONOMER   \n",
      "4619            FUCULOKIN-MONOMER   \n",
      "4620              FUCISOM-MONOMER   \n",
      "4621            FUCPALDOL-MONOMER   \n",
      "4622  FRUCTBISALD-CLASSII-MONOMER   \n",
      "4623              EG10144-MONOMER   \n",
      "4624              EG10114-MONOMER   \n",
      "4625              EG10085-MONOMER   \n",
      "4626           RIBULPEPIM-MONOMER   \n",
      "4627            RIBULOKIN-MONOMER   \n",
      "4628             ARABISOM-MONOMER   \n",
      "4629             KDGALDOL-MONOMER   \n",
      "\n",
      "                          GO terms (molecular function)  \\\n",
      "0                                                   NaN   \n",
      "1                                                   NaN   \n",
      "2                                                   NaN   \n",
      "3                                                   NaN   \n",
      "4                                                   NaN   \n",
      "5                                                   NaN   \n",
      "6                                            GO:0008565   \n",
      "7                                                   NaN   \n",
      "8                                                   NaN   \n",
      "9                              GO:0003677 // GO:0000150   \n",
      "10                                                  NaN   \n",
      "11                                                  NaN   \n",
      "12                                                  NaN   \n",
      "13               GO:0004803 // GO:0003677 // GO:0003676   \n",
      "14                                           GO:0016740   \n",
      "15                                                  NaN   \n",
      "16                             GO:0016810 // GO:0016787   \n",
      "17                                           GO:0008452   \n",
      "18                                                  NaN   \n",
      "19                                                  NaN   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                                  NaN   \n",
      "25                                                  NaN   \n",
      "26                                                  NaN   \n",
      "27                                                  NaN   \n",
      "28                                                  NaN   \n",
      "29                                                  NaN   \n",
      "...                                                 ...   \n",
      "4600  GO:0043798 // GO:0016740 // GO:0005524 // GO:0...   \n",
      "4601  GO:0016787 // GO:0042803 // GO:0034338 // GO:0...   \n",
      "4602  GO:0016773 // GO:0016740 // GO:0016301 // GO:0...   \n",
      "4603  GO:0016853 // GO:0000287 // GO:0046872 // GO:0...   \n",
      "4604  GO:0016829 // GO:0030145 // GO:0008198 // GO:0...   \n",
      "4605  GO:0050662 // GO:0016491 // GO:0051287 // GO:0...   \n",
      "4606  GO:0042802 // GO:0016798 // GO:0016787 // GO:0...   \n",
      "4607             GO:0042802 // GO:0016491 // GO:0009010   \n",
      "4608  GO:0046872 // GO:0042803 // GO:0016773 // GO:0...   \n",
      "4609  GO:0042802 // GO:0016853 // GO:0048029 // GO:0...   \n",
      "4610  GO:0046872 // GO:0042802 // GO:0042803 // GO:0...   \n",
      "4611  GO:0046872 // GO:0016740 // GO:0016301 // GO:0...   \n",
      "4612  GO:0050662 // GO:0003824 // GO:0016491 // GO:0...   \n",
      "4613  GO:0003824 // GO:0046872 // GO:0016798 // GO:0...   \n",
      "4614                                                NaN   \n",
      "4615  GO:0046872 // GO:0016798 // GO:0016787 // GO:0...   \n",
      "4616  GO:0016407 // GO:0016740 // GO:0016746 // GO:0...   \n",
      "4617  GO:0016798 // GO:0016787 // GO:0004553 // GO:0...   \n",
      "4618  GO:0048029 // GO:0016854 // GO:0016853 // GO:0...   \n",
      "4619  GO:0016773 // GO:0016740 // GO:0016301 // GO:0...   \n",
      "4620  GO:0042802 // GO:0016861 // GO:0016853 // GO:0...   \n",
      "4621  GO:0016829 // GO:0008270 // GO:0046872 // GO:0...   \n",
      "4622  GO:0042803 // GO:0016832 // GO:0016829 // GO:0...   \n",
      "4623  GO:0003824 // GO:0016798 // GO:0016787 // GO:0...   \n",
      "4624  GO:0008422 // GO:0004553 // GO:0016798 // GO:0...   \n",
      "4625  GO:0008422 // GO:0004553 // GO:0016798 // GO:0...   \n",
      "4626  GO:0016832 // GO:0016853 // GO:0008270 // GO:0...   \n",
      "4627  GO:0016773 // GO:0016740 // GO:0016301 // GO:0...   \n",
      "4628  GO:0016861 // GO:0016853 // GO:0030145 // GO:0...   \n",
      "4629  GO:0016829 // GO:0003824 // GO:0016830 // GO:0...   \n",
      "\n",
      "                          GO terms (biological process)  \\\n",
      "0                                                   NaN   \n",
      "1                                            GO:0042710   \n",
      "2                                                   NaN   \n",
      "3                                                   NaN   \n",
      "4                                                   NaN   \n",
      "5                                                   NaN   \n",
      "6                              GO:0015031 // GO:0006810   \n",
      "7                                                   NaN   \n",
      "8                                                   NaN   \n",
      "9                              GO:0015074 // GO:0006310   \n",
      "10                                                  NaN   \n",
      "11                                                  NaN   \n",
      "12    GO:0075713 // GO:0046718 // GO:0015074 // GO:0...   \n",
      "13                                           GO:0006313   \n",
      "14                                           GO:0008152   \n",
      "15                                           GO:0097264   \n",
      "16                                           GO:0006807   \n",
      "17                                           GO:0006396   \n",
      "18                                                  NaN   \n",
      "19                                           GO:0009306   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                                  NaN   \n",
      "25                                                  NaN   \n",
      "26                                                  NaN   \n",
      "27                                                  NaN   \n",
      "28                                                  NaN   \n",
      "29                                                  NaN   \n",
      "...                                                 ...   \n",
      "4600  GO:0031388 // GO:0016310 // GO:0046392 // GO:0...   \n",
      "4601  GO:0009056 // GO:0008152 // GO:0043433 // GO:0...   \n",
      "4602  GO:0046835 // GO:0042732 // GO:0005997 // GO:0...   \n",
      "4603  GO:0042732 // GO:0005975 // GO:0006098 // GO:0...   \n",
      "4604             GO:0006064 // GO:0006974 // GO:0042840   \n",
      "4605                           GO:0055114 // GO:0019698   \n",
      "4606  GO:0051289 // GO:0008152 // GO:0005975 // GO:0...   \n",
      "4607  GO:0051289 // GO:0008152 // GO:0006062 // GO:0...   \n",
      "4608  GO:0046835 // GO:0006014 // GO:0005975 // GO:0...   \n",
      "4609             GO:0005996 // GO:0005975 // GO:0019303   \n",
      "4610  GO:0061615 // GO:0005975 // GO:0006974 // GO:0...   \n",
      "4611  GO:0061615 // GO:0051289 // GO:0046835 // GO:0...   \n",
      "4612  GO:0019592 // GO:0019594 // GO:0008152 // GO:0...   \n",
      "4613  GO:0008152 // GO:0005975 // GO:0055114 // GO:0...   \n",
      "4614                           GO:0008643 // GO:0006810   \n",
      "4615             GO:0008152 // GO:0005975 // GO:0005990   \n",
      "4616                                         GO:0005989   \n",
      "4617  GO:0006974 // GO:0005980 // GO:0008152 // GO:0...   \n",
      "4618  GO:0042354 // GO:0005996 // GO:0006004 // GO:0...   \n",
      "4619  GO:0046835 // GO:0019317 // GO:0006004 // GO:0...   \n",
      "4620  GO:0019317 // GO:0006004 // GO:0005996 // GO:0...   \n",
      "4621  GO:0019317 // GO:0006004 // GO:0005975 // GO:0...   \n",
      "4622             GO:0006094 // GO:0005975 // GO:0006096   \n",
      "4623  GO:0008152 // GO:0005975 // GO:0055114 // GO:0...   \n",
      "4624  GO:1901657 // GO:0044724 // GO:0005975 // GO:0...   \n",
      "4625  GO:1901657 // GO:0044724 // GO:2000892 // GO:0...   \n",
      "4626  GO:0019572 // GO:0019568 // GO:0005975 // GO:0...   \n",
      "4627  GO:0046835 // GO:0016052 // GO:0005975 // GO:0...   \n",
      "4628  GO:0008152 // GO:0005996 // GO:0005975 // GO:0...   \n",
      "4629             GO:0019394 // GO:0046392 // GO:0042838   \n",
      "\n",
      "                                    Transcription Units  \\\n",
      "0                                             TU0-14204   \n",
      "1                                             TU0-14204   \n",
      "2                                             TU0-14184   \n",
      "3                                               TU00201   \n",
      "4                                             TU0-14035   \n",
      "5                                             TU0-13940   \n",
      "6                                             TU0-13811   \n",
      "7                                             TU0-13767   \n",
      "8                                             TU0-13765   \n",
      "9                                             TU0-13701   \n",
      "10                                            TU0-13602   \n",
      "11                                             TU0-7744   \n",
      "12                                             TU0-8838   \n",
      "13                                            TU0-13184   \n",
      "14                                            TU0-13118   \n",
      "15                                            TU0-13027   \n",
      "16                                            TU0-13002   \n",
      "17                                            TU0-12855   \n",
      "18                                            TU0-12852   \n",
      "19                                            TU0-14247   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                            TU0-14186   \n",
      "25                                            TU0-14016   \n",
      "26                                            TU0-13950   \n",
      "27                                            TU0-13564   \n",
      "28                                            TU0-13332   \n",
      "29                                            TU0-13155   \n",
      "...                                                 ...   \n",
      "4600                                              TU482   \n",
      "4601                                           TU0-5421   \n",
      "4602                                               TU67   \n",
      "4603                                               TU67   \n",
      "4604                                              TU101   \n",
      "4605                               TU0-36422 // TU00372   \n",
      "4606                                           TU0-5081   \n",
      "4607                                            TU00133   \n",
      "4608                                            TU00206   \n",
      "4609                                            TU00206   \n",
      "4610                               TU0-8181 // TU0-6408   \n",
      "4611                                 TU0-14813 // TU788   \n",
      "4612                                            TU00193   \n",
      "4613                                            TU00131   \n",
      "4614                                            TU00123   \n",
      "4615        TU0-4703 // TU0-4702 // TU0-4701 // TU00036   \n",
      "4616  TU0-14521 // TU0-4703 // TU0-4702 // TU0-4701 ...   \n",
      "4617                                          TU0-14302   \n",
      "4618                                            TU00128   \n",
      "4619                                            TU00128   \n",
      "4620                                            TU00128   \n",
      "4621                                            TU00135   \n",
      "4622     TU0-14818 // TU0-14816 // TU0-8082 // TU0-8081   \n",
      "4623                                            TU00248   \n",
      "4624                                           TU0-4802   \n",
      "4625                               TU0-42599 // TU00241   \n",
      "4626                                            TU00214   \n",
      "4627                                            TU00214   \n",
      "4628                                            TU00214   \n",
      "4629                                              TU482   \n",
      "\n",
      "     Transcription-Direction  \\\n",
      "0                          +   \n",
      "1                          +   \n",
      "2                          +   \n",
      "3                          -   \n",
      "4                          -   \n",
      "5                          +   \n",
      "6                          -   \n",
      "7                          -   \n",
      "8                          +   \n",
      "9                          -   \n",
      "10                         +   \n",
      "11                         -   \n",
      "12                         +   \n",
      "13                         -   \n",
      "14                         +   \n",
      "15                         +   \n",
      "16                         +   \n",
      "17                         +   \n",
      "18                         +   \n",
      "19                         -   \n",
      "20                         +   \n",
      "21                         -   \n",
      "22                         -   \n",
      "23                         +   \n",
      "24                         -   \n",
      "25                         -   \n",
      "26                         +   \n",
      "27                         +   \n",
      "28                         -   \n",
      "29                         +   \n",
      "...                      ...   \n",
      "4600                       -   \n",
      "4601                       -   \n",
      "4602                       -   \n",
      "4603                       -   \n",
      "4604                       +   \n",
      "4605                       -   \n",
      "4606                       -   \n",
      "4607                       +   \n",
      "4608                       +   \n",
      "4609                       +   \n",
      "4610                       +   \n",
      "4611                       +   \n",
      "4612                       +   \n",
      "4613                       +   \n",
      "4614                       +   \n",
      "4615                       -   \n",
      "4616                       -   \n",
      "4617                       -   \n",
      "4618                       +   \n",
      "4619                       +   \n",
      "4620                       +   \n",
      "4621                       -   \n",
      "4622                       -   \n",
      "4623                       -   \n",
      "4624                       -   \n",
      "4625                       +   \n",
      "4626                       -   \n",
      "4627                       -   \n",
      "4628                       -   \n",
      "4629                       -   \n",
      "\n",
      "                            Genes of transcription unit  \\\n",
      "0                                        G7935 // G7934   \n",
      "1                                        G7935 // G7934   \n",
      "2                                                 G7903   \n",
      "3     EG10725 // EG10724 // EG10723 // EG10722 // EG...   \n",
      "4                                        G7798 // G7797   \n",
      "5          G7762 // G7761 // G0-10540 // G7760 // G7759   \n",
      "6                                        G7536 // G7535   \n",
      "7     G7481 // G7480 // G0-10544 // G7479 // G7478 /...   \n",
      "8                                                 G7474   \n",
      "9                                                 G7383   \n",
      "10                                                G7222   \n",
      "11                                       G7120 // G7122   \n",
      "12                                                G7044   \n",
      "13                                     G6632 // EG11611   \n",
      "14                                       G6536 // G6535   \n",
      "15                                     EG11523 // G6376   \n",
      "16                                     G6348 // EG12208   \n",
      "17                                     EG11496 // G6119   \n",
      "18                                                G6114   \n",
      "19                                                G6113   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                    G0-10567 // G7905   \n",
      "25            EG11710 // EG11709 // EG11708 // G0-10564   \n",
      "26                                             G0-10563   \n",
      "27                                  G0-10561 // EG12323   \n",
      "28                                             G0-10559   \n",
      "29                                             G0-10556   \n",
      "...                                                 ...   \n",
      "4600  EG30069 // EG11175 // EG11176 // EG10016 // EG...   \n",
      "4601                                            EG11101   \n",
      "4602                                 EG11075 // EG11074   \n",
      "4603                                 EG11075 // EG11074   \n",
      "4604                                 EG20248 // EG11066   \n",
      "4605                                            EG11065   \n",
      "4606                        G6866 // EG11658 // EG11055   \n",
      "4607  G8210 // EG10969 // EG10970 // EG10971 // EG10...   \n",
      "4608  EG10819 // EG10817 // EG10814 // EG10816 // EG...   \n",
      "4609  EG10819 // EG10817 // EG10814 // EG10816 // EG...   \n",
      "4610                                            EG10700   \n",
      "4611                                            EG10699   \n",
      "4612                      EG10615 // EG10616 // EG11965   \n",
      "4613                                 EG10577 // EG10578   \n",
      "4614                      EG10558 // EG10528 // EG10559   \n",
      "4615                      EG10524 // EG10526 // EG10527   \n",
      "4616                      EG10524 // EG10526 // EG10527   \n",
      "4617  EG10380 // EG10377 // EG10379 // EG10381 // EG...   \n",
      "4618  EG10352 // EG10349 // EG10350 // EG10355 // EG...   \n",
      "4619  EG10352 // EG10349 // EG10350 // EG10355 // EG...   \n",
      "4620  EG10352 // EG10349 // EG10350 // EG10355 // EG...   \n",
      "4621                                 EG10351 // EG10348   \n",
      "4622                      EG10282 // EG10703 // EG10368   \n",
      "4623  EG12198 // EG10144 // EG10143 // EG10142 // EG...   \n",
      "4624                      EG10114 // EG10115 // EG10116   \n",
      "4625                                 EG10085 // EG10086   \n",
      "4626                      EG10055 // EG10052 // EG10053   \n",
      "4627                      EG10055 // EG10052 // EG10053   \n",
      "4628                      EG10055 // EG10052 // EG10053   \n",
      "4629  EG30069 // EG11175 // EG11176 // EG10016 // EG...   \n",
      "\n",
      "               Transcription units - promoter of operon  \\\n",
      "0                                                   NaN   \n",
      "1                                                   NaN   \n",
      "2                                                   NaN   \n",
      "3                                               PM00284   \n",
      "4                                                   NaN   \n",
      "5                                                   NaN   \n",
      "6                                                   NaN   \n",
      "7                                                   NaN   \n",
      "8                                                   NaN   \n",
      "9                                                   NaN   \n",
      "10                                                  NaN   \n",
      "11                                            PM0-10331   \n",
      "12                                                  NaN   \n",
      "13                                                  NaN   \n",
      "14                                                  NaN   \n",
      "15                                                  NaN   \n",
      "16                                                  NaN   \n",
      "17                                                  NaN   \n",
      "18                                                  NaN   \n",
      "19                                                  NaN   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                                  NaN   \n",
      "25                                                  NaN   \n",
      "26                                                  NaN   \n",
      "27                                                  NaN   \n",
      "28                                                  NaN   \n",
      "29                                                  NaN   \n",
      "...                                                 ...   \n",
      "4600                                              PM582   \n",
      "4601                                           PM0-5661   \n",
      "4602                                               PM68   \n",
      "4603                                               PM68   \n",
      "4604                                              PM101   \n",
      "4605                                    PM0-36188 // \"\"   \n",
      "4606                                           PM0-5321   \n",
      "4607                                            PM00178   \n",
      "4608                                            PM00290   \n",
      "4609                                            PM00290   \n",
      "4610                                PM0-8561 // PM00377   \n",
      "4611                                 PM0-10550 // PM933   \n",
      "4612                                            PM00272   \n",
      "4613                                            PM00175   \n",
      "4614                                            PM00048   \n",
      "4615        PM0-4943 // PM0-4942 // PM0-4941 // PM00045   \n",
      "4616  PM0-9804 // PM0-4943 // PM0-4942 // PM0-4941 /...   \n",
      "4617                                          PM0-36176   \n",
      "4618                                            PM00172   \n",
      "4619                                            PM00172   \n",
      "4620                                            PM00172   \n",
      "4621                                            PM00180   \n",
      "4622      PM0-10554 // PM0-10551 // PM00591 // PM0-8461   \n",
      "4623                                              PM183   \n",
      "4624                                           PM0-5042   \n",
      "4625                              PM0-46347 // PM0-8848   \n",
      "4626                                            PM00003   \n",
      "4627                                            PM00003   \n",
      "4628                                            PM00003   \n",
      "4629                                              PM582   \n",
      "\n",
      "                                Sequence - DNA sequence  \n",
      "0     ATGCTATTTTATCACGGTAGCCGTAGTCCTTACCCCTGGTCCCTGT...  \n",
      "1     ATGACAAACTTCACGACCAGCACGCCGCATGATGCATTATTTAAAA...  \n",
      "2     TTGCTGGCCGATGAGCTGACCATTGGGCCTATCCGGGCTGTCCCGA...  \n",
      "3     GTGCTGGCGCTGTTTATCCACACCACCGGCGTGCTCTCCAAGCTGC...  \n",
      "4     GTGGCGATTAATTTCAGCCCTAAGGTTGGTGAAATACTGGAATGCA...  \n",
      "5     ATGACTAAAACCCAAATAAATAAATTAATAAAAATGATGAATGATT...  \n",
      "6     GTGATTCACCAACAGCATATGCGCAACATTGCGCAGTGGTTGCAGG...  \n",
      "7     ATGCGAAAAAAAATAGAGATGAGCTTGATAAAATCTCCTGCAAATG...  \n",
      "8     ATGTTCAGAATCGAACCAAATTTAATTAAAGCTATCGCTCTTGTTG...  \n",
      "9     ATGTGGCATCTTGTCGTGCTGTTGGAGGAGCTGTGCGAACGTGGTA...  \n",
      "10    ATGCGTCCCGCGACTATCGCCCCATTAACGCCATACGATAAATGGG...  \n",
      "11    GCCCGTGTAGCAGAGTTCTCCATCAGCCTGGCTACCGGTCGGGCAG...  \n",
      "12    GTGGTCGTAACAACATCCGACGTGGTGATGTGTCAAATGAGACGCT...  \n",
      "13    ATGCGACCGTTCGCCGCCGTCGTTTACCGGGGGACATGGTTATCTG...  \n",
      "14    ATGCGCGATTATCTTGCCGGACTGGATATCGCTTCGCCAGAGCATC...  \n",
      "15    ATGTGGCCGGATAACCGTATCGCCCGTGACGCGCACTATCTTTACC...  \n",
      "16    ATGATGACGACAATTCTGACGATTCATGTTCCTTCAACGCCGGGGC...  \n",
      "17    ATGGAATGGTACATGGGCAAATATATTCGTCCCTTATCCGATGCGG...  \n",
      "18    GTCAGTAAATCCGAGCGTGAAAGCATTATCGCCGCGCTGCACGGGC...  \n",
      "19    CTGTTGACAATCAACTTTATTGTCGTCACCAAAGGGGCCGAGCGTA...  \n",
      "20    GATTTCTCACTGACAGATAAAGAAATGGCACAAATTAGTTCGCTTG...  \n",
      "21    ATGACACAGGCAGAACGCCGCCATGACCGGCTGGCTGTCAGGCTGT...  \n",
      "22    ATTATCGCCATTGTTGTCATCGTCATCATTCTGAACAAAACCGGGG...  \n",
      "23                       AACTTTGATTTATAGTCAGGTGGGGCTTTT  \n",
      "24    ATGAGCAGAAAAACCCAACGTTACTCTAAAAAGTTCAAAGCCGAAG...  \n",
      "25    ATGAGTCAGCAAGTGGCGGTGGAAAAACTGGTCGTGGACGCCTGGG...  \n",
      "26    ATGTCAAATACATACCAGAAAAGAAAGGCAAGTAAAGAGTATGGTT...  \n",
      "27    ATGACTATTGCTGAACGGCTGCGGCAGGAAGGACATCAAATTGGCT...  \n",
      "28    ATGACAATAGAAAAACACGAAAGAAGCACTAAGGATTTGGTGAAAG...  \n",
      "29    ATGGGTTGGACTGAAGCGGCTGACCTGATTGTTAAAGGTATGGAAG...  \n",
      "...                                                 ...  \n",
      "4600  ATGAAAATCGTAATCGCCCCAGACTCTTATAAAGAAAGTTTATCTG...  \n",
      "4601  ATGAAGCCGGAAAACAAACTACCTGTTCTGGACCTTATTTCTGCTG...  \n",
      "4602  ATGTATATCGGGATAGATCTTGGCACCTCGGGCGTAAAAGTTATTT...  \n",
      "4603  ATGCAAGCCTATTTTGACCAGCTCGATCGCGTTCGTTATGAAGGCT...  \n",
      "4604  ATGGAACAGACCTGGCGCTGGTACGGCCCAAACGATCCGGTTTCTT...  \n",
      "4605  GTGAAAACACTAAATCGTCGCGATTTTCCCGGTGCACAGTATCCAG...  \n",
      "4606  ATGTTACGTCCTGTAGAAACCCCAACCCGTGAAATCAAAAAACTCG...  \n",
      "4607  ATGAATCAGGTTGCCGTTGTCATCGGTGGTGGGCAAACCTTAGGCG...  \n",
      "4608  ATGCAAAACGCAGGCAGCCTCGTTGTTCTTGGCAGCATTAATGCTG...  \n",
      "4609  ATGAAAAAAGGCACCGTTCTTAATTCTGATATTTCATCGGTGATCT...  \n",
      "4610  ATGGTACGTATCTATACGTTGACACTTGCGCCCTCTCTCGATAGCG...  \n",
      "4611  ATGATTAAGAAAATCGGTGTGTTGACAAGCGGCGGTGATGCGCCAG...  \n",
      "4612  ATGAAAGCATTACATTTTGGCGCAGGTAATATCGGTCGTGGCTTTA...  \n",
      "4613  ATGATGTCTGCACCCAAAATTACATTTATCGGCGCTGGTTCGACGA...  \n",
      "4614  ATGAAAATGAATAAAAGTCTCATCGTCCTCTGTTTATCAGCAGGGT...  \n",
      "4615  ATGACCATGATTACGGATTCACTGGCCGTCGTTTTACAACGTCGTG...  \n",
      "4616  TTGAACATGCCAATGACCGAAAGAATAAGAGCAGGCAAGCTATTTA...  \n",
      "4617  ATGACACAACTCGCCATTGGCAAACCCGCTCCCCTCGGCGCGCATT...  \n",
      "4618  ATGCTGAAAACAATTTCGCCGTTAATTTCTCCCGAACTATTGAAAG...  \n",
      "4619  ATGTTATCCGGCTATATTGCAGGAGCGATTATGAAACAAGAAGTTA...  \n",
      "4620  ATGAAAAAAATCAGCTTACCGAAAATTGGTATCCGCCCGGTTATTG...  \n",
      "4621  ATGGAACGAAATAAACTTGCTCGTCAGATTATTGACACTTGCCTGG...  \n",
      "4622  ATGTCTAAGATTTTTGATTTCGTAAAACCTGGCGTAATCACTGGTG...  \n",
      "4623  ATGAGCCAGAAATTAAAAGTCGTCACTATTGGTGGCGGGAGCAGCT...  \n",
      "4624  ATGAAAGCATTTCCAGAAACATTTCTTTGGGGTGGCGCAACAGCTG...  \n",
      "4625  ATGTCAGTATTTCCAGAAAGTTTTTTATGGGGCGGCGCGCTTGCCG...  \n",
      "4626  ATGTTAGAAGATCTCAAACGCCAGGTATTAGAAGCCAACCTGGCGC...  \n",
      "4627  ATGGCGATTGCAATTGGCCTCGATTTTGGCAGTGATTCTGTGCGAG...  \n",
      "4628  ATGACGATTTTTGATAATTATGAAGTGTGGTTTGTCATTGGCAGCC...  \n",
      "4629  ATGAATAACGATGTTTTCCCGAATAAATTCAAAGCCGCACTGGCTG...  \n",
      "\n",
      "[4630 rows x 12 columns]\n"
     ]
    }
   ],
   "source": [
    "#instantiate an organism BB for ecoli\n",
    "#myOrg=organismBreadBoard('ECOLI')\n",
    "#myOrg=pc.select_organism('ecoli')\n",
    "#instantiate a fasta reader with target genome\n",
    "myReader = FastAreader ('ecoli_v3_u00096_3.fa')\n",
    "myDFrame1=pd.DataFrame(columns=['Gene','Strand','Sequence','Promoter','Transcription Units','Other genes in operon','Protospacers'])\n",
    "\n",
    "#parse fastareader and pull out the header and sequence for the genome\n",
    "targetHead=''\n",
    "targetSeq=''\n",
    "for head, seq in myReader.readFasta() :\n",
    "    targetHead+=head\n",
    "    targetSeq+= seq\n",
    "myTable_genes = pd.read_table('All_genes_of_E._coli_K-12_substr._MG1655_10172016_2.txt')\n",
    "print myTable_genes\n",
    "for i, row in myTable_genes.iterrows():\n",
    "    myDFrame1.loc[i,['Gene']]=row['Gene Name']\n",
    "    myDFrame1.loc[i,['Strand']]=row['Transcription-Direction']\n",
    "    myDFrame1.loc[i,['Promoter']]=row['Transcription units - promoter of operon']\n",
    "    myDFrame1.loc[i, ['Transcription Units']]=row['Transcription Units']\n",
    "    myDFrame1.loc[i, ['Other genes in operon']]=row['Genes of transcription unit']\n",
    "    myDFrame1.loc[i, ['Sequence']]=row['Sequence - DNA sequence']\n",
    "    #if not math.isnan(row['Left-End-Position']):\n",
    "        #geneSeq+=targetSeq[int(row['Left-End-Position']): int(row['Right-End-Position'])]\n",
    "    #else:\n",
    "        #continue\n",
    "    #myDFrame1.loc[i,['Sequence']]=geneSeq\n",
    "#myDFrame1.to_csv('genes_promos_Ecoli', sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "          Gene Strand                                           Sequence  \\\n",
      "0        G7935      +  ATGCTATTTTATCACGGTAGCCGTAGTCCTTACCCCTGGTCCCTGT...   \n",
      "1        G7934      +  ATGACAAACTTCACGACCAGCACGCCGCATGATGCATTATTTAAAA...   \n",
      "2        G7903      +  TTGCTGGCCGATGAGCTGACCATTGGGCCTATCCGGGCTGTCCCGA...   \n",
      "3        G7822      -  GTGCTGGCGCTGTTTATCCACACCACCGGCGTGCTCTCCAAGCTGC...   \n",
      "4        G7798      -  GTGGCGATTAATTTCAGCCCTAAGGTTGGTGAAATACTGGAATGCA...   \n",
      "5        G7760      +  ATGACTAAAACCCAAATAAATAAATTAATAAAAATGATGAATGATT...   \n",
      "6        G7536      -  GTGATTCACCAACAGCATATGCGCAACATTGCGCAGTGGTTGCAGG...   \n",
      "7        G7478      -  ATGCGAAAAAAAATAGAGATGAGCTTGATAAAATCTCCTGCAAATG...   \n",
      "8        G7474      +  ATGTTCAGAATCGAACCAAATTTAATTAAAGCTATCGCTCTTGTTG...   \n",
      "9        G7383      -  ATGTGGCATCTTGTCGTGCTGTTGGAGGAGCTGTGCGAACGTGGTA...   \n",
      "10       G7222      +  ATGCGTCCCGCGACTATCGCCCCATTAACGCCATACGATAAATGGG...   \n",
      "11       G7120      -  GCCCGTGTAGCAGAGTTCTCCATCAGCCTGGCTACCGGTCGGGCAG...   \n",
      "12       G7044      +  GTGGTCGTAACAACATCCGACGTGGTGATGTGTCAAATGAGACGCT...   \n",
      "13       G6632      -  ATGCGACCGTTCGCCGCCGTCGTTTACCGGGGGACATGGTTATCTG...   \n",
      "14       G6535      +  ATGCGCGATTATCTTGCCGGACTGGATATCGCTTCGCCAGAGCATC...   \n",
      "15       G6376      +  ATGTGGCCGGATAACCGTATCGCCCGTGACGCGCACTATCTTTACC...   \n",
      "16       G6348      +  ATGATGACGACAATTCTGACGATTCATGTTCCTTCAACGCCGGGGC...   \n",
      "17       G6119      +  ATGGAATGGTACATGGGCAAATATATTCGTCCCTTATCCGATGCGG...   \n",
      "18       G6114      +  GTCAGTAAATCCGAGCGTGAAAGCATTATCGCCGCGCTGCACGGGC...   \n",
      "19       G6113      -  CTGTTGACAATCAACTTTATTGTCGTCACCAAAGGGGCCGAGCGTA...   \n",
      "20    G0-10682      +  GATTTCTCACTGACAGATAAAGAAATGGCACAAATTAGTTCGCTTG...   \n",
      "21    G0-10615      -  ATGACACAGGCAGAACGCCGCCATGACCGGCTGGCTGTCAGGCTGT...   \n",
      "22    G0-10614      -  ATTATCGCCATTGTTGTCATCGTCATCATTCTGAACAAAACCGGGG...   \n",
      "23    G0-10604      +                     AACTTTGATTTATAGTCAGGTGGGGCTTTT   \n",
      "24    G0-10567      -  ATGAGCAGAAAAACCCAACGTTACTCTAAAAAGTTCAAAGCCGAAG...   \n",
      "25    G0-10564      -  ATGAGTCAGCAAGTGGCGGTGGAAAAACTGGTCGTGGACGCCTGGG...   \n",
      "26    G0-10563      +  ATGTCAAATACATACCAGAAAAGAAAGGCAAGTAAAGAGTATGGTT...   \n",
      "27    G0-10561      +  ATGACTATTGCTGAACGGCTGCGGCAGGAAGGACATCAAATTGGCT...   \n",
      "28    G0-10559      -  ATGACAATAGAAAAACACGAAAGAAGCACTAAGGATTTGGTGAAAG...   \n",
      "29    G0-10556      +  ATGGGTTGGACTGAAGCGGCTGACCTGATTGTTAAAGGTATGGAAG...   \n",
      "...        ...    ...                                                ...   \n",
      "4600   EG11175      -  ATGAAAATCGTAATCGCCCCAGACTCTTATAAAGAAAGTTTATCTG...   \n",
      "4601   EG11101      -  ATGAAGCCGGAAAACAAACTACCTGTTCTGGACCTTATTTCTGCTG...   \n",
      "4602   EG11075      -  ATGTATATCGGGATAGATCTTGGCACCTCGGGCGTAAAAGTTATTT...   \n",
      "4603   EG11074      -  ATGCAAGCCTATTTTGACCAGCTCGATCGCGTTCGTTATGAAGGCT...   \n",
      "4604   EG11066      +  ATGGAACAGACCTGGCGCTGGTACGGCCCAAACGATCCGGTTTCTT...   \n",
      "4605   EG11065      -  GTGAAAACACTAAATCGTCGCGATTTTCCCGGTGCACAGTATCCAG...   \n",
      "4606   EG11055      -  ATGTTACGTCCTGTAGAAACCCCAACCCGTGAAATCAAAAAACTCG...   \n",
      "4607   EG10971      +  ATGAATCAGGTTGCCGTTGTCATCGGTGGTGGGCAAACCTTAGGCG...   \n",
      "4608   EG10818      +  ATGCAAAACGCAGGCAGCCTCGTTGTTCTTGGCAGCATTAATGCTG...   \n",
      "4609   EG10817      +  ATGAAAAAAGGCACCGTTCTTAATTCTGATATTTCATCGGTGATCT...   \n",
      "4610   EG10700      +  ATGGTACGTATCTATACGTTGACACTTGCGCCCTCTCTCGATAGCG...   \n",
      "4611   EG10699      +  ATGATTAAGAAAATCGGTGTGTTGACAAGCGGCGGTGATGCGCCAG...   \n",
      "4612   EG10616      +  ATGAAAGCATTACATTTTGGCGCAGGTAATATCGGTCGTGGCTTTA...   \n",
      "4613   EG10577      +  ATGATGTCTGCACCCAAAATTACATTTATCGGCGCTGGTTCGACGA...   \n",
      "4614   EG10559      +  ATGAAAATGAATAAAAGTCTCATCGTCCTCTGTTTATCAGCAGGGT...   \n",
      "4615   EG10527      -  ATGACCATGATTACGGATTCACTGGCCGTCGTTTTACAACGTCGTG...   \n",
      "4616   EG10524      -  TTGAACATGCCAATGACCGAAAGAATAAGAGCAGGCAAGCTATTTA...   \n",
      "4617   EG10381      -  ATGACACAACTCGCCATTGGCAAACCCGCTCCCCTCGGCGCGCATT...   \n",
      "4618   EG10355      +  ATGCTGAAAACAATTTCGCCGTTAATTTCTCCCGAACTATTGAAAG...   \n",
      "4619   EG10350      +  ATGTTATCCGGCTATATTGCAGGAGCGATTATGAAACAAGAAGTTA...   \n",
      "4620   EG10349      +  ATGAAAAAAATCAGCTTACCGAAAATTGGTATCCGCCCGGTTATTG...   \n",
      "4621   EG10348      -  ATGGAACGAAATAAACTTGCTCGTCAGATTATTGACACTTGCCTGG...   \n",
      "4622   EG10282      -  ATGTCTAAGATTTTTGATTTCGTAAAACCTGGCGTAATCACTGGTG...   \n",
      "4623   EG10144      -  ATGAGCCAGAAATTAAAAGTCGTCACTATTGGTGGCGGGAGCAGCT...   \n",
      "4624   EG10114      -  ATGAAAGCATTTCCAGAAACATTTCTTTGGGGTGGCGCAACAGCTG...   \n",
      "4625   EG10085      +  ATGTCAGTATTTCCAGAAAGTTTTTTATGGGGCGGCGCGCTTGCCG...   \n",
      "4626   EG10055      -  ATGTTAGAAGATCTCAAACGCCAGGTATTAGAAGCCAACCTGGCGC...   \n",
      "4627   EG10053      -  ATGGCGATTGCAATTGGCCTCGATTTTGGCAGTGATTCTGTGCGAG...   \n",
      "4628   EG10052      -  ATGACGATTTTTGATAATTATGAAGTGTGGTTTGTCATTGGCAGCC...   \n",
      "4629   EG10016      -  ATGAATAACGATGTTTTCCCGAATAAATTCAAAGCCGCACTGGCTG...   \n",
      "\n",
      "                                               Promoter  \\\n",
      "0                                                   NaN   \n",
      "1                                                   NaN   \n",
      "2                                                   NaN   \n",
      "3                                               PM00284   \n",
      "4                                                   NaN   \n",
      "5                                                   NaN   \n",
      "6                                                   NaN   \n",
      "7                                                   NaN   \n",
      "8                                                   NaN   \n",
      "9                                                   NaN   \n",
      "10                                                  NaN   \n",
      "11                                            PM0-10331   \n",
      "12                                                  NaN   \n",
      "13                                                  NaN   \n",
      "14                                                  NaN   \n",
      "15                                                  NaN   \n",
      "16                                                  NaN   \n",
      "17                                                  NaN   \n",
      "18                                                  NaN   \n",
      "19                                                  NaN   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                                  NaN   \n",
      "25                                                  NaN   \n",
      "26                                                  NaN   \n",
      "27                                                  NaN   \n",
      "28                                                  NaN   \n",
      "29                                                  NaN   \n",
      "...                                                 ...   \n",
      "4600                                              PM582   \n",
      "4601                                           PM0-5661   \n",
      "4602                                               PM68   \n",
      "4603                                               PM68   \n",
      "4604                                              PM101   \n",
      "4605                                    PM0-36188 // \"\"   \n",
      "4606                                           PM0-5321   \n",
      "4607                                            PM00178   \n",
      "4608                                            PM00290   \n",
      "4609                                            PM00290   \n",
      "4610                                PM0-8561 // PM00377   \n",
      "4611                                 PM0-10550 // PM933   \n",
      "4612                                            PM00272   \n",
      "4613                                            PM00175   \n",
      "4614                                            PM00048   \n",
      "4615        PM0-4943 // PM0-4942 // PM0-4941 // PM00045   \n",
      "4616  PM0-9804 // PM0-4943 // PM0-4942 // PM0-4941 /...   \n",
      "4617                                          PM0-36176   \n",
      "4618                                            PM00172   \n",
      "4619                                            PM00172   \n",
      "4620                                            PM00172   \n",
      "4621                                            PM00180   \n",
      "4622      PM0-10554 // PM0-10551 // PM00591 // PM0-8461   \n",
      "4623                                              PM183   \n",
      "4624                                           PM0-5042   \n",
      "4625                              PM0-46347 // PM0-8848   \n",
      "4626                                            PM00003   \n",
      "4627                                            PM00003   \n",
      "4628                                            PM00003   \n",
      "4629                                              PM582   \n",
      "\n",
      "                                    Transcription Units  \\\n",
      "0                                             TU0-14204   \n",
      "1                                             TU0-14204   \n",
      "2                                             TU0-14184   \n",
      "3                                               TU00201   \n",
      "4                                             TU0-14035   \n",
      "5                                             TU0-13940   \n",
      "6                                             TU0-13811   \n",
      "7                                             TU0-13767   \n",
      "8                                             TU0-13765   \n",
      "9                                             TU0-13701   \n",
      "10                                            TU0-13602   \n",
      "11                                             TU0-7744   \n",
      "12                                             TU0-8838   \n",
      "13                                            TU0-13184   \n",
      "14                                            TU0-13118   \n",
      "15                                            TU0-13027   \n",
      "16                                            TU0-13002   \n",
      "17                                            TU0-12855   \n",
      "18                                            TU0-12852   \n",
      "19                                            TU0-14247   \n",
      "20                                                  NaN   \n",
      "21                                                  NaN   \n",
      "22                                                  NaN   \n",
      "23                                                  NaN   \n",
      "24                                            TU0-14186   \n",
      "25                                            TU0-14016   \n",
      "26                                            TU0-13950   \n",
      "27                                            TU0-13564   \n",
      "28                                            TU0-13332   \n",
      "29                                            TU0-13155   \n",
      "...                                                 ...   \n",
      "4600                                              TU482   \n",
      "4601                                           TU0-5421   \n",
      "4602                                               TU67   \n",
      "4603                                               TU67   \n",
      "4604                                              TU101   \n",
      "4605                               TU0-36422 // TU00372   \n",
      "4606                                           TU0-5081   \n",
      "4607                                            TU00133   \n",
      "4608                                            TU00206   \n",
      "4609                                            TU00206   \n",
      "4610                               TU0-8181 // TU0-6408   \n",
      "4611                                 TU0-14813 // TU788   \n",
      "4612                                            TU00193   \n",
      "4613                                            TU00131   \n",
      "4614                                            TU00123   \n",
      "4615        TU0-4703 // TU0-4702 // TU0-4701 // TU00036   \n",
      "4616  TU0-14521 // TU0-4703 // TU0-4702 // TU0-4701 ...   \n",
      "4617                                          TU0-14302   \n",
      "4618                                            TU00128   \n",
      "4619                                            TU00128   \n",
      "4620                                            TU00128   \n",
      "4621                                            TU00135   \n",
      "4622     TU0-14818 // TU0-14816 // TU0-8082 // TU0-8081   \n",
      "4623                                            TU00248   \n",
      "4624                                           TU0-4802   \n",
      "4625                               TU0-42599 // TU00241   \n",
      "4626                                            TU00214   \n",
      "4627                                            TU00214   \n",
      "4628                                            TU00214   \n",
      "4629                                              TU482   \n",
      "\n",
      "                                  Other genes in operon Protospacers  \n",
      "0                                        G7935 // G7934          NaN  \n",
      "1                                        G7935 // G7934          NaN  \n",
      "2                                                 G7903          NaN  \n",
      "3     EG10725 // EG10724 // EG10723 // EG10722 // EG...          NaN  \n",
      "4                                        G7798 // G7797          NaN  \n",
      "5          G7762 // G7761 // G0-10540 // G7760 // G7759          NaN  \n",
      "6                                        G7536 // G7535          NaN  \n",
      "7     G7481 // G7480 // G0-10544 // G7479 // G7478 /...          NaN  \n",
      "8                                                 G7474          NaN  \n",
      "9                                                 G7383          NaN  \n",
      "10                                                G7222          NaN  \n",
      "11                                       G7120 // G7122          NaN  \n",
      "12                                                G7044          NaN  \n",
      "13                                     G6632 // EG11611          NaN  \n",
      "14                                       G6536 // G6535          NaN  \n",
      "15                                     EG11523 // G6376          NaN  \n",
      "16                                     G6348 // EG12208          NaN  \n",
      "17                                     EG11496 // G6119          NaN  \n",
      "18                                                G6114          NaN  \n",
      "19                                                G6113          NaN  \n",
      "20                                                  NaN          NaN  \n",
      "21                                                  NaN          NaN  \n",
      "22                                                  NaN          NaN  \n",
      "23                                                  NaN          NaN  \n",
      "24                                    G0-10567 // G7905          NaN  \n",
      "25            EG11710 // EG11709 // EG11708 // G0-10564          NaN  \n",
      "26                                             G0-10563          NaN  \n",
      "27                                  G0-10561 // EG12323          NaN  \n",
      "28                                             G0-10559          NaN  \n",
      "29                                             G0-10556          NaN  \n",
      "...                                                 ...          ...  \n",
      "4600  EG30069 // EG11175 // EG11176 // EG10016 // EG...          NaN  \n",
      "4601                                            EG11101          NaN  \n",
      "4602                                 EG11075 // EG11074          NaN  \n",
      "4603                                 EG11075 // EG11074          NaN  \n",
      "4604                                 EG20248 // EG11066          NaN  \n",
      "4605                                            EG11065          NaN  \n",
      "4606                        G6866 // EG11658 // EG11055          NaN  \n",
      "4607  G8210 // EG10969 // EG10970 // EG10971 // EG10...          NaN  \n",
      "4608  EG10819 // EG10817 // EG10814 // EG10816 // EG...          NaN  \n",
      "4609  EG10819 // EG10817 // EG10814 // EG10816 // EG...          NaN  \n",
      "4610                                            EG10700          NaN  \n",
      "4611                                            EG10699          NaN  \n",
      "4612                      EG10615 // EG10616 // EG11965          NaN  \n",
      "4613                                 EG10577 // EG10578          NaN  \n",
      "4614                      EG10558 // EG10528 // EG10559          NaN  \n",
      "4615                      EG10524 // EG10526 // EG10527          NaN  \n",
      "4616                      EG10524 // EG10526 // EG10527          NaN  \n",
      "4617  EG10380 // EG10377 // EG10379 // EG10381 // EG...          NaN  \n",
      "4618  EG10352 // EG10349 // EG10350 // EG10355 // EG...          NaN  \n",
      "4619  EG10352 // EG10349 // EG10350 // EG10355 // EG...          NaN  \n",
      "4620  EG10352 // EG10349 // EG10350 // EG10355 // EG...          NaN  \n",
      "4621                                 EG10351 // EG10348          NaN  \n",
      "4622                      EG10282 // EG10703 // EG10368          NaN  \n",
      "4623  EG12198 // EG10144 // EG10143 // EG10142 // EG...          NaN  \n",
      "4624                      EG10114 // EG10115 // EG10116          NaN  \n",
      "4625                                 EG10085 // EG10086          NaN  \n",
      "4626                      EG10055 // EG10052 // EG10053          NaN  \n",
      "4627                      EG10055 // EG10052 // EG10053          NaN  \n",
      "4628                      EG10055 // EG10052 // EG10053          NaN  \n",
      "4629  EG30069 // EG11175 // EG11176 // EG10016 // EG...          NaN  \n",
      "\n",
      "[4630 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "print myDFrame1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "myTable_promoters = pd.read_table('All_promoters_of_E._coli_K-12_substr._MG1655_10172016.txt')\n",
    "#print myTable_promoters.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<bound method DataFrame.head of      Promoter Name  Absolute-Plus-1-Position  \\\n",
      "0        PM0-45785                 3604315.0   \n",
      "1        PM0-10222                  850209.0   \n",
      "2        PM0-46187                 1686654.0   \n",
      "3         PM0-4102                   17317.0   \n",
      "4         PM0-9365                  498031.0   \n",
      "5        PM0-44676                  960143.0   \n",
      "6          PM00083                 1122643.0   \n",
      "7        PM0-45040                 1764924.0   \n",
      "8          PM00538                 2798556.0   \n",
      "9        PM0-45404                 2667928.0   \n",
      "10       PM0-45769                 3577085.0   \n",
      "11       PM0-10194                 4478445.0   \n",
      "12       PM0-46159                 1563201.0   \n",
      "13        PM0-3664                 1890561.0   \n",
      "14        PM0-9308                  403687.0   \n",
      "15       PM0-44660                  922750.0   \n",
      "16         PM00058                  703708.0   \n",
      "17       PM0-45024                 1739855.0   \n",
      "18         PM00522                  953925.0   \n",
      "19       PM0-45388                 2629672.0   \n",
      "20       PM0-45753                 3526319.0   \n",
      "21       PM0-10177                 4264283.0   \n",
      "22       PM0-46123                 1344510.0   \n",
      "23       PM0-36248                 2590970.0   \n",
      "24        PM0-9061                 4556856.0   \n",
      "25       PM0-44644                  862783.0   \n",
      "26         PM00037                 2352583.0   \n",
      "27       PM0-45008                 1692988.0   \n",
      "28         PM00505                 4483919.0   \n",
      "29       PM0-45372                 2558693.0   \n",
      "...            ...                       ...   \n",
      "3811     PM0-46324                 1413556.0   \n",
      "3812     PM0-44362                  148920.0   \n",
      "3813      PM0-9560                 3990790.0   \n",
      "3814       PM00147                 1686709.0   \n",
      "3815     PM0-44727                 1105714.0   \n",
      "3816     PM0-45091                 1879390.0   \n",
      "3817       PM00591                 3073823.0   \n",
      "3818     PM0-45455                 2843102.0   \n",
      "3819     PM0-45820                 3696283.0   \n",
      "3820     PM0-10299                 2797085.0   \n",
      "3821     PM0-46306                 3382173.0   \n",
      "3822     PM0-44348                   51293.0   \n",
      "3823      PM0-9521                  401319.0   \n",
      "3824     PM0-44711                 1056222.0   \n",
      "3825       PM00132                 3305860.0   \n",
      "3826     PM0-45075                 1854041.0   \n",
      "3827       PM00574                 2850839.0   \n",
      "3828     PM0-45439                 2758855.0   \n",
      "3829     PM0-45804                 3626848.0   \n",
      "3830     PM0-10254                  883569.0   \n",
      "3831     PM0-46268                 3648294.0   \n",
      "3832      PM0-4433                 3201104.0   \n",
      "3833      PM0-9408                 3247548.0   \n",
      "3834     PM0-44695                 1030061.0   \n",
      "3835       PM00107                 2891964.0   \n",
      "3836     PM0-45059                 1813348.0   \n",
      "3837       PM00557                   10643.0   \n",
      "3838     PM0-45423                 2741355.0   \n",
      "3839     PM0-45788                 3605575.0   \n",
      "3840     PM0-10228                 3055974.0   \n",
      "\n",
      "               Promoters - genes downstream of promoter  \\\n",
      "0                                                   NaN   \n",
      "1                                               EG12117   \n",
      "2                                               EG10358   \n",
      "3                                    EG11078 // EG10652   \n",
      "4                                               EG10431   \n",
      "5                                                   NaN   \n",
      "6                                               EG10806   \n",
      "7                                                   NaN   \n",
      "8                                               EG11554   \n",
      "9                                                   NaN   \n",
      "10                                                  NaN   \n",
      "11                                                G7885   \n",
      "12    G6777 // G6778 // G6779 // G6780 // G6781 // G...   \n",
      "13                                                G6990   \n",
      "14                                              EG11257   \n",
      "15                                                  NaN   \n",
      "16             EG10634 // EG10636 // EG10632 // EG10633   \n",
      "17                                                  NaN   \n",
      "18                                              EG10701   \n",
      "19                                                  NaN   \n",
      "20                                                  NaN   \n",
      "21                                              EG10236   \n",
      "22                                                G6638   \n",
      "23                                              G0-8888   \n",
      "24                                                G7923   \n",
      "25                                                  NaN   \n",
      "26                        EG10391 // EG10392 // EG10393   \n",
      "27                                                  NaN   \n",
      "28                                              EG11067   \n",
      "29                                                  NaN   \n",
      "...                                                 ...   \n",
      "3811                                           G0-16600   \n",
      "3812                                                NaN   \n",
      "3813                                            EG10170   \n",
      "3814                                            EG10358   \n",
      "3815                                                NaN   \n",
      "3816                                                NaN   \n",
      "3817  EG10282 // EG10703 // EG10368 // EG10703 // EG...   \n",
      "3818                                                NaN   \n",
      "3819                                                NaN   \n",
      "3820                            G7395 // G7396 // G7396   \n",
      "3821                                            EG11652   \n",
      "3822                                                NaN   \n",
      "3823                                            EG11256   \n",
      "3824                                                NaN   \n",
      "3825                                            EG10617   \n",
      "3826                                                NaN   \n",
      "3827  EG10301 // EG10487 // EG10486 // EG10485 // EG...   \n",
      "3828                                                NaN   \n",
      "3829                                                NaN   \n",
      "3830                                              G6440   \n",
      "3831                                           G0-10752   \n",
      "3832                                            EG12434   \n",
      "3833                                     G7610 // G7609   \n",
      "3834                                                NaN   \n",
      "3835                      EG10189 // EG10190 // EG10191   \n",
      "3836                                                NaN   \n",
      "3837                                            EG11509   \n",
      "3838                                                NaN   \n",
      "3839                                                NaN   \n",
      "3840                      EG30099 // EG11158 // EG30099   \n",
      "\n",
      "     Transcription-Direction Promoters - transcription units of promoter  \\\n",
      "0                          +                                         NaN   \n",
      "1                        NaN                                   TU0-14688   \n",
      "2                        NaN                                   TU0-42482   \n",
      "3                        NaN                                    TU0-3862   \n",
      "4                        NaN                                   TU0-12951   \n",
      "5                          +                                         NaN   \n",
      "6                        NaN                                     TU00060   \n",
      "7                          -                                         NaN   \n",
      "8                        NaN                                     TU00394   \n",
      "9                          -                                         NaN   \n",
      "10                         -                                         NaN   \n",
      "11                       NaN                                    TU0-8617   \n",
      "12                       NaN                                   TU0-42462   \n",
      "13                       NaN                                    TU0-3464   \n",
      "14                         +                                    TU0-8872   \n",
      "15                         +                                         NaN   \n",
      "16                       NaN                                     TU00043   \n",
      "17                         +                                         NaN   \n",
      "18                       NaN                                     TU00384   \n",
      "19                         +                                         NaN   \n",
      "20                         +                                         NaN   \n",
      "21                       NaN                                   TU0-14119   \n",
      "22                       NaN                                   TU0-42448   \n",
      "23                       NaN                                   TU0-14237   \n",
      "24                       NaN                                    TU0-8625   \n",
      "25                         -                                         NaN   \n",
      "26                       NaN                                     TU00030   \n",
      "27                         -                                         NaN   \n",
      "28                       NaN                                    TU0-6687   \n",
      "29                         +                                         NaN   \n",
      "...                      ...                                         ...   \n",
      "3811                       +                                   TU0-42572   \n",
      "3812                       -                                         NaN   \n",
      "3813                     NaN                                   TU0-14402   \n",
      "3814                     NaN                                     TU00105   \n",
      "3815                       +                                         NaN   \n",
      "3816                       -                                         NaN   \n",
      "3817                     NaN                         TU0-8082 // TU00443   \n",
      "3818                       -                                         NaN   \n",
      "3819                       -                                         NaN   \n",
      "3820                       -                      TU0-42608 // TU0-13709   \n",
      "3821                       +                                   TU0-42561   \n",
      "3822                       -                                         NaN   \n",
      "3823                     NaN                                   TU0-12913   \n",
      "3824                       +                                         NaN   \n",
      "3825                     NaN                                     TU00093   \n",
      "3826                       -                                         NaN   \n",
      "3827                     NaN                         TU0-8345 // TU00424   \n",
      "3828                       +                                         NaN   \n",
      "3829                       -                                         NaN   \n",
      "3830                     NaN                                   TU0-14699   \n",
      "3831                     NaN                                   TU0-42530   \n",
      "3832                     NaN                                    TU0-4192   \n",
      "3833                     NaN                                   TU0-14286   \n",
      "3834                       +                                         NaN   \n",
      "3835                     NaN                                     TU00078   \n",
      "3836                       -                                         NaN   \n",
      "3837                     NaN                                     TU00406   \n",
      "3838                       +                                         NaN   \n",
      "3839                       +                                         NaN   \n",
      "3840                     NaN                      TU0-36424 // TU0-13794   \n",
      "\n",
      "      Transcription units - pathways  \n",
      "0                                NaN  \n",
      "1                                NaN  \n",
      "2                                NaN  \n",
      "3                                NaN  \n",
      "4                                NaN  \n",
      "5                                NaN  \n",
      "6                                NaN  \n",
      "7                                NaN  \n",
      "8                                NaN  \n",
      "9                                NaN  \n",
      "10                               NaN  \n",
      "11                               NaN  \n",
      "12                               NaN  \n",
      "13                               NaN  \n",
      "14                               NaN  \n",
      "15                               NaN  \n",
      "16                               NaN  \n",
      "17                               NaN  \n",
      "18                               NaN  \n",
      "19                               NaN  \n",
      "20                               NaN  \n",
      "21                               NaN  \n",
      "22                               NaN  \n",
      "23                               NaN  \n",
      "24                               NaN  \n",
      "25                               NaN  \n",
      "26                               NaN  \n",
      "27                               NaN  \n",
      "28                               NaN  \n",
      "29                               NaN  \n",
      "...                              ...  \n",
      "3811                             NaN  \n",
      "3812                             NaN  \n",
      "3813                             NaN  \n",
      "3814                             NaN  \n",
      "3815                             NaN  \n",
      "3816                             NaN  \n",
      "3817                             NaN  \n",
      "3818                             NaN  \n",
      "3819                             NaN  \n",
      "3820                             NaN  \n",
      "3821                             NaN  \n",
      "3822                             NaN  \n",
      "3823                             NaN  \n",
      "3824                             NaN  \n",
      "3825                             NaN  \n",
      "3826                             NaN  \n",
      "3827                             NaN  \n",
      "3828                             NaN  \n",
      "3829                             NaN  \n",
      "3830                             NaN  \n",
      "3831                             NaN  \n",
      "3832                             NaN  \n",
      "3833                             NaN  \n",
      "3834                             NaN  \n",
      "3835                             NaN  \n",
      "3836                             NaN  \n",
      "3837                             NaN  \n",
      "3838                             NaN  \n",
      "3839                             NaN  \n",
      "3840                             NaN  \n",
      "\n",
      "[3841 rows x 6 columns]>\n"
     ]
    }
   ],
   "source": [
    "myTable_promoters = pd.read_table('All_promoters_of_E._coli_K-12_substr._MG1655_10172016.txt')\n",
    "myDFrame2=pd.DataFrame(columns=['Promoter','Strand','Position','Sequence','Genes in operon','Transcription Units','Pathways'])\n",
    "upstream = 60\n",
    "downstream = 60\n",
    "for i, row in myTable_promoters.iterrows():\n",
    "    myDFrame2.loc[i,['Promoter']]=row['Promoter Name']\n",
    "    myDFrame2.loc[i, ['Strand']]=row['Transcription-Direction']\n",
    "    myDFrame2.loc[i,['Position']]=row['Absolute-Plus-1-Position']\n",
    "    myDFrame2.loc[i,['Genes in operon']]=row['Promoters - genes downstream of promoter']\n",
    "    myDFrame2.loc[i, ['Transcription Units']]=row['Promoters - transcription units of promoter']\n",
    "    myDFrame2.loc[i, ['Genes in operon']]=row['Promoters - genes downstream of promoter']\n",
    "    myDFrame2.loc[i, ['Pathways']]=row['Transcription units - pathways']\n",
    "    if not math.isnan(row['Absolute-Plus-1-Position']):\n",
    "        if row['Absolute-Plus-1-Position']!='-':\n",
    "            myDFrame2.loc[i, ['Sequence']]=targetSeq[(int(row['Absolute-Plus-1-Position'])-upstream): (int(row['Absolute-Plus-1-Position']+downstream))]\n",
    "        else:\n",
    "            myDFrame2.loc[i, ['Sequence']]=targetSeq[(int(row['Absolute-Plus-1-Position'])-downstream): (int(row['Absolute-Plus-1-Position']+upstream))]\n",
    "    else:\n",
    "        #print row['Promoter Name']\n",
    "        #print row['Absolute-Plus-1-Position']\n",
    "        #myDFrame2.loc[i, ['Sequence']]=targetSeq[(int(row['Absolute-Plus-1-Position'])-55): (int(row['Absolute-Plus-1-Position']+30))]\n",
    "        continue\n",
    "print myTable_promoters.head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       Promoter Strand     Position  \\\n",
      "0     PM0-45785      +  3.60432e+06   \n",
      "1     PM0-10222    NaN       850209   \n",
      "2     PM0-46187    NaN  1.68665e+06   \n",
      "3      PM0-4102    NaN        17317   \n",
      "4      PM0-9365    NaN       498031   \n",
      "5     PM0-44676      +       960143   \n",
      "6       PM00083    NaN  1.12264e+06   \n",
      "7     PM0-45040      -  1.76492e+06   \n",
      "8       PM00538    NaN  2.79856e+06   \n",
      "9     PM0-45404      -  2.66793e+06   \n",
      "10    PM0-45769      -  3.57708e+06   \n",
      "11    PM0-10194    NaN  4.47844e+06   \n",
      "12    PM0-46159    NaN   1.5632e+06   \n",
      "13     PM0-3664    NaN  1.89056e+06   \n",
      "14     PM0-9308      +       403687   \n",
      "15    PM0-44660      +       922750   \n",
      "16      PM00058    NaN       703708   \n",
      "17    PM0-45024      +  1.73986e+06   \n",
      "18      PM00522    NaN       953925   \n",
      "19    PM0-45388      +  2.62967e+06   \n",
      "20    PM0-45753      +  3.52632e+06   \n",
      "21    PM0-10177    NaN  4.26428e+06   \n",
      "22    PM0-46123    NaN  1.34451e+06   \n",
      "23    PM0-36248    NaN  2.59097e+06   \n",
      "24     PM0-9061    NaN  4.55686e+06   \n",
      "25    PM0-44644      -       862783   \n",
      "26      PM00037    NaN  2.35258e+06   \n",
      "27    PM0-45008      -  1.69299e+06   \n",
      "28      PM00505    NaN  4.48392e+06   \n",
      "29    PM0-45372      +  2.55869e+06   \n",
      "...         ...    ...          ...   \n",
      "3811  PM0-46324      +  1.41356e+06   \n",
      "3812  PM0-44362      -       148920   \n",
      "3813   PM0-9560    NaN  3.99079e+06   \n",
      "3814    PM00147    NaN  1.68671e+06   \n",
      "3815  PM0-44727      +  1.10571e+06   \n",
      "3816  PM0-45091      -  1.87939e+06   \n",
      "3817    PM00591    NaN  3.07382e+06   \n",
      "3818  PM0-45455      -   2.8431e+06   \n",
      "3819  PM0-45820      -  3.69628e+06   \n",
      "3820  PM0-10299      -  2.79708e+06   \n",
      "3821  PM0-46306      +  3.38217e+06   \n",
      "3822  PM0-44348      -        51293   \n",
      "3823   PM0-9521    NaN       401319   \n",
      "3824  PM0-44711      +  1.05622e+06   \n",
      "3825    PM00132    NaN  3.30586e+06   \n",
      "3826  PM0-45075      -  1.85404e+06   \n",
      "3827    PM00574    NaN  2.85084e+06   \n",
      "3828  PM0-45439      +  2.75886e+06   \n",
      "3829  PM0-45804      -  3.62685e+06   \n",
      "3830  PM0-10254    NaN       883569   \n",
      "3831  PM0-46268    NaN  3.64829e+06   \n",
      "3832   PM0-4433    NaN   3.2011e+06   \n",
      "3833   PM0-9408    NaN  3.24755e+06   \n",
      "3834  PM0-44695      +  1.03006e+06   \n",
      "3835    PM00107    NaN  2.89196e+06   \n",
      "3836  PM0-45059      -  1.81335e+06   \n",
      "3837    PM00557    NaN        10643   \n",
      "3838  PM0-45423      +  2.74136e+06   \n",
      "3839  PM0-45788      +  3.60558e+06   \n",
      "3840  PM0-10228    NaN  3.05597e+06   \n",
      "\n",
      "                                               Sequence  \\\n",
      "0     TCCTCGCTGTTGCTTCATGGCACAGCGTTAACGCTATGTACATAGC...   \n",
      "1     TACCCAATGCTGGTTGAGCATTTGTTGAAAAAATTTTCCCCCGTTT...   \n",
      "2     CTCCTCACCTGATTAATTTTTTCTTTCTGTTTTGCTTTCGTTAAGC...   \n",
      "3     ACTTTTCCTAATTAATCCCTCAGGAATCCTCACCTTAAGCTATGAT...   \n",
      "4     GATTTGACCTCTCACAGCAATTAGTTCTTCTTCCTCACTTTTCCGC...   \n",
      "5     CAGGCAGCCTGAATGAACAACGGGCAATAAATAGCCAAATCTTTCT...   \n",
      "6     TAATACCTGGGATGGTGCAGTCATTAATCTCTATGCTCCGGCTGAA...   \n",
      "7     AGAAGGTCATCGAACCTGGACGGAGGTTAATCCAGGTCGATTTGGC...   \n",
      "8     ATAAAGAAAACCTTTTAACGCCAAAACAAAAAAAGTATTTCGTCGC...   \n",
      "9     TTAGTCTGGGGTTAATCGAATGTTGGTGCAAGTTGAAAAATAAGAA...   \n",
      "10    CCCTACTGCTGTTAAAACAAGCGTTAATGTAGTCACTACTTATTTG...   \n",
      "11    AAGCATAAATCTTTGTCACAAAGGTGGAGGCAATGTCAGTGGTGTG...   \n",
      "12    GGATAAGCGCCAGAATGCTGGCTTAAAGTTATATTTTTTAACTTTT...   \n",
      "13    TGCCTTTGATAACATTCTTTTGAACCGCCATGGTGCGACTCCTTTT...   \n",
      "14    ACGAAAATGGCTCACCCGAAATATCATACTTCTGCCTTTAGCTCCG...   \n",
      "15    GAGAATTAGGGCGAGCCGTCAAGCATTTGACAGGGGACAAGGGGCA...   \n",
      "16    ATTAGGCATCTTTAAGCGTAACCTGGATTTGCGCAGACAGGCGTCA...   \n",
      "17    CATAGTCTGCCTGCAAAATTTTTGAAACCAGTCATCAAATATTACC...   \n",
      "18    TCGCGACCGGCAGCACCATAATGAACGCTTTGTCCATCAGGCTGCG...   \n",
      "19    CTGTTTGGCTGCGAAATAAATATAAAATTAATATATATGTTGTAAT...   \n",
      "20    CAAATCAAATTAAAAGTCAATAGGTTGAAATAACTCCAGGAATTTG...   \n",
      "21    GTTGCCATGTGTCCTCCAGAACGTATCGTCAGGGTCTGCTTCATAT...   \n",
      "22    TGCCTTCGGGATGTCTGGCGTAGATTCAGCAACCTGGCGCAAGAGT...   \n",
      "23    CCGTGGTGAAGCTAATTTACTCGATTGCTGCGATGACTACCTCCGG...   \n",
      "24    ACTTTTTTCCTGAAAAGTCAGTAGATATTCATAGGCAAGTAAGGTT...   \n",
      "25    AAGTGATATTTGTTGATGCCCAGCGTGTGGTAGGGCAGAAAATGAA...   \n",
      "26    ATGTTCAAAATGACGCATGAAATCACGTTTCACTTTCGAATTATGA...   \n",
      "27    TTTACGATTATCAATTTCAACCACGATTTCTTTGAATTTTTTATCC...   \n",
      "28    GGTTGAAACCAGCCACGCGGTAGGCTTTATAGCGTTCGCGCGCCAG...   \n",
      "29    CCTTAAACTGATATACATTGCCGAACAATGTGGTCTTTAGTTTCAT...   \n",
      "...                                                 ...   \n",
      "3811  CCGATGTTGATTATTAATAATCAGCTATGAAGTTTTAATTTGAATA...   \n",
      "3812  ATACTGCCGCACAGCCGCGCGGATGTCGCCCGTTTCGGCGAGGAAA...   \n",
      "3813  TTGGCGTGTGGCAATTCTTAAAACATTGTCTAACATGCTTGTTACC...   \n",
      "3814  CTCACTTATTATTTACCATTTGATAACAAATGTTTGGTCTTTCGTG...   \n",
      "3815  TAGCCTCAGTAAAAGTGCAGTATCTTGATAAACAGAAAAAGCTGGC...   \n",
      "3816  GATTTGTGACGAGCATCACGGAGGTTTTTTTTGCGATGGCGCAGAA...   \n",
      "3817  TACCTGTCGCAAACTGCGACTGATTGGTTAATTGTCGAACATTTAA...   \n",
      "3818  CGCAGTCTCCTCACCATCAAATAGATGGCGTTTTCCTGATGAATAT...   \n",
      "3819  GACAATTCTAACCGTGAGCATTTCGCCGGTTTATTGTTTACGGCGA...   \n",
      "3820  CATATGTACTCCCTATAAGAAAATTACTCATTGTTTAAAAAGAGAT...   \n",
      "3821  CGTAAACCGGGCATCAGGCTTACGTGTGATGTCCGGTTAACTCGTG...   \n",
      "3822  GAATGAGTGTGGGAACGGCGAGTCGGAATACGGGAATGTCGATGCT...   \n",
      "3823  CTGGTAATCAAACAAAAAATATTTGCGCAAAGTATTTCCTTTGTCA...   \n",
      "3824  TCAGGGTTAAATTCACGGTCGGTGCACTTTAGGTGAAAAAGGTTGA...   \n",
      "3825  GTGGTTAGTGTTGCCATGAGGGCTTCTCTCCAGTGAAAAATAGTGC...   \n",
      "3826  GGTGCCTCGGATTTGCTAAAAAGGTAGCAATTGCGGCAAATTCACT...   \n",
      "3827  CCTTTTGTTTTGATCTGGTTTGCCGCGGCAGCGTGGCGGAAGGTTG...   \n",
      "3828  TGTCGGATGGATTGAGGCCGAGATTGATGAGTGGTTATCACAACGC...   \n",
      "3829  GCGTAAATGGCGCATCATCCCTCCTGTTTTTTCAGCAGCAGGATAC...   \n",
      "3830  GGTGATTGTAACGGGCATAGGTTAAATAAAACTTAAAGAAAGCGTA...   \n",
      "3831  CCAGCTGAAGAAATCGCTAATTCTTGCAATGTTAGCCACTGGCTAA...   \n",
      "3832  TTTTGCGCCAAATTGCCATGCAACGGGCAATTTGACGGGCGTAAAA...   \n",
      "3833  AAAGTGTCAGCCTGTGTAAATCCTCTCGCCACCCTCCCCTGCATTC...   \n",
      "3834  CGTCATTGTCCGGAGTTTTCCATGGATTTGCAACGCGTCCATTAAG...   \n",
      "3835  TTATGTTCCAGTAAGCAAAGCTGTTTCTGCGCCCTGTCAGCGCCCA...   \n",
      "3836  AAAAGATCACTTATTTATAGCAATAGATCGTCAAAGGCAGCTTTTT...   \n",
      "3837  GTGGGGCAGTGTCTACGATCTTTTGAGGGGAAAATGAAAATTTTCC...   \n",
      "3838  CTGTATAGATAAATTTACACTCCCTTTGAAAACAATCCGCTATGCT...   \n",
      "3839  TCAGCTTTTTCCCGTCGCTGTTGCATAAAACGAAACGATGCGGCGA...   \n",
      "3840  GAAAAAACTAACCAAAACTTTGAATGACACTTTTCGGTTTACTGTG...   \n",
      "\n",
      "                                        Genes in operon  \\\n",
      "0                                                   NaN   \n",
      "1                                               EG12117   \n",
      "2                                               EG10358   \n",
      "3                                    EG11078 // EG10652   \n",
      "4                                               EG10431   \n",
      "5                                                   NaN   \n",
      "6                                               EG10806   \n",
      "7                                                   NaN   \n",
      "8                                               EG11554   \n",
      "9                                                   NaN   \n",
      "10                                                  NaN   \n",
      "11                                                G7885   \n",
      "12    G6777 // G6778 // G6779 // G6780 // G6781 // G...   \n",
      "13                                                G6990   \n",
      "14                                              EG11257   \n",
      "15                                                  NaN   \n",
      "16             EG10634 // EG10636 // EG10632 // EG10633   \n",
      "17                                                  NaN   \n",
      "18                                              EG10701   \n",
      "19                                                  NaN   \n",
      "20                                                  NaN   \n",
      "21                                              EG10236   \n",
      "22                                                G6638   \n",
      "23                                              G0-8888   \n",
      "24                                                G7923   \n",
      "25                                                  NaN   \n",
      "26                        EG10391 // EG10392 // EG10393   \n",
      "27                                                  NaN   \n",
      "28                                              EG11067   \n",
      "29                                                  NaN   \n",
      "...                                                 ...   \n",
      "3811                                           G0-16600   \n",
      "3812                                                NaN   \n",
      "3813                                            EG10170   \n",
      "3814                                            EG10358   \n",
      "3815                                                NaN   \n",
      "3816                                                NaN   \n",
      "3817  EG10282 // EG10703 // EG10368 // EG10703 // EG...   \n",
      "3818                                                NaN   \n",
      "3819                                                NaN   \n",
      "3820                            G7395 // G7396 // G7396   \n",
      "3821                                            EG11652   \n",
      "3822                                                NaN   \n",
      "3823                                            EG11256   \n",
      "3824                                                NaN   \n",
      "3825                                            EG10617   \n",
      "3826                                                NaN   \n",
      "3827  EG10301 // EG10487 // EG10486 // EG10485 // EG...   \n",
      "3828                                                NaN   \n",
      "3829                                                NaN   \n",
      "3830                                              G6440   \n",
      "3831                                           G0-10752   \n",
      "3832                                            EG12434   \n",
      "3833                                     G7610 // G7609   \n",
      "3834                                                NaN   \n",
      "3835                      EG10189 // EG10190 // EG10191   \n",
      "3836                                                NaN   \n",
      "3837                                            EG11509   \n",
      "3838                                                NaN   \n",
      "3839                                                NaN   \n",
      "3840                      EG30099 // EG11158 // EG30099   \n",
      "\n",
      "         Transcription Units Pathways  \n",
      "0                        NaN      NaN  \n",
      "1                  TU0-14688      NaN  \n",
      "2                  TU0-42482      NaN  \n",
      "3                   TU0-3862      NaN  \n",
      "4                  TU0-12951      NaN  \n",
      "5                        NaN      NaN  \n",
      "6                    TU00060      NaN  \n",
      "7                        NaN      NaN  \n",
      "8                    TU00394      NaN  \n",
      "9                        NaN      NaN  \n",
      "10                       NaN      NaN  \n",
      "11                  TU0-8617      NaN  \n",
      "12                 TU0-42462      NaN  \n",
      "13                  TU0-3464      NaN  \n",
      "14                  TU0-8872      NaN  \n",
      "15                       NaN      NaN  \n",
      "16                   TU00043      NaN  \n",
      "17                       NaN      NaN  \n",
      "18                   TU00384      NaN  \n",
      "19                       NaN      NaN  \n",
      "20                       NaN      NaN  \n",
      "21                 TU0-14119      NaN  \n",
      "22                 TU0-42448      NaN  \n",
      "23                 TU0-14237      NaN  \n",
      "24                  TU0-8625      NaN  \n",
      "25                       NaN      NaN  \n",
      "26                   TU00030      NaN  \n",
      "27                       NaN      NaN  \n",
      "28                  TU0-6687      NaN  \n",
      "29                       NaN      NaN  \n",
      "...                      ...      ...  \n",
      "3811               TU0-42572      NaN  \n",
      "3812                     NaN      NaN  \n",
      "3813               TU0-14402      NaN  \n",
      "3814                 TU00105      NaN  \n",
      "3815                     NaN      NaN  \n",
      "3816                     NaN      NaN  \n",
      "3817     TU0-8082 // TU00443      NaN  \n",
      "3818                     NaN      NaN  \n",
      "3819                     NaN      NaN  \n",
      "3820  TU0-42608 // TU0-13709      NaN  \n",
      "3821               TU0-42561      NaN  \n",
      "3822                     NaN      NaN  \n",
      "3823               TU0-12913      NaN  \n",
      "3824                     NaN      NaN  \n",
      "3825                 TU00093      NaN  \n",
      "3826                     NaN      NaN  \n",
      "3827     TU0-8345 // TU00424      NaN  \n",
      "3828                     NaN      NaN  \n",
      "3829                     NaN      NaN  \n",
      "3830               TU0-14699      NaN  \n",
      "3831               TU0-42530      NaN  \n",
      "3832                TU0-4192      NaN  \n",
      "3833               TU0-14286      NaN  \n",
      "3834                     NaN      NaN  \n",
      "3835                 TU00078      NaN  \n",
      "3836                     NaN      NaN  \n",
      "3837                 TU00406      NaN  \n",
      "3838                     NaN      NaN  \n",
      "3839                     NaN      NaN  \n",
      "3840  TU0-36424 // TU0-13794      NaN  \n",
      "\n",
      "[3841 rows x 7 columns]\n"
     ]
    }
   ],
   "source": [
    "print myDFrame2\n",
    "myDFrame2.to_csv('test_out.CSV')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#Makes protospacers from an arbitrary sequence (in our case either a gene or promoter)\n",
    "\n",
    "#Note, ignores promoter ids that don't have associated promoter; need to add those.\n",
    "#Note: Often the algorithm below cannot pull a full protospacer_length protospacer from the given \"promoter\" region; in this case\n",
    "#nothing is pulled and an empty comma shows up. Need to update this.\n",
    "\n",
    "#Currently not adding lists of protospacers directly to dataframe (myDFrame2)-- need to add this capability.\n",
    "\n",
    "protospacer_length = 30;\n",
    "PAM_length = 3;\n",
    "myDFrame2['protospacer-positive'] = \"\"\n",
    "myDFrame2['protospacer-negative'] = \"\"\n",
    "protospacers = []\n",
    "for i in range(myDFrame2.shape[0]):\n",
    "    targetSeq =  myDFrame2.loc[i, ['Sequence']][0]\n",
    "    protospacers_i = []\n",
    "    if(type(targetSeq) == str):\n",
    "        for base in range(len(targetSeq)):\n",
    "            if (targetSeq[base-2:base] == 'GG') & (base-(protospacer_length+PAM_length)>-1):\n",
    "                protospacer_positive = targetSeq[base-(protospacer_length+PAM_length):base-PAM_length];\n",
    "                myDFrame2.loc[i, ['protospacer-positive']] += protospacer_positive + \", \" \n",
    "                protospacers_i.append(protospacer_positive)\n",
    "            elif (targetSeq[base-3:base-1] == 'CC') & (base+(protospacer_length)<len(targetSeq)):\n",
    "                protospacer_negative = targetSeq[base:base+(protospacer_length)];\n",
    "                myDFrame2.loc[i, ['protospacer-negative']] += protospacer_negative + \", \"\n",
    "                protospacers_i.append(protospacer_negative)\n",
    "            else:\n",
    "                continue\n",
    "    protospacers += protospacers_i\n",
    "                \n",
    "myDFrame2.to_csv('CRISPRiTargets.CSV')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Target Sequence: GCGCGCTTTAAAGCCTCCAGTGCGGTGTTTAAATCTTTGTGGGATCAGGGCATTATCTTACGTGATCAGAATAAACAACCCTCTTTAAGCGGCTGCCTGCGAATTACCGTCGGAACCCGT\n",
      "\n",
      "negative strand\n",
      "PAM: CCT\n",
      "Protospacer: CCAGTGCGGTGTTTAAATCTTTGTGGGATC\n",
      "\n",
      "negative strand\n",
      "PAM: CCA\n",
      "Protospacer: GTGCGGTGTTTAAATCTTTGTGGGATCAGG\n",
      "\n",
      "positive strand\n",
      "PAM: TGG\n",
      "Protospacer: AAAGCCTCCAGTGCGGTGTTTAAATCTTTG\n",
      "\n",
      "positive strand\n",
      "PAM: GGG\n",
      "Protospacer: AAGCCTCCAGTGCGGTGTTTAAATCTTTGT\n",
      "\n",
      "positive strand\n",
      "PAM: AGG\n",
      "Protospacer: CCAGTGCGGTGTTTAAATCTTTGTGGGATC\n",
      "\n",
      "positive strand\n",
      "PAM: GGG\n",
      "Protospacer: CAGTGCGGTGTTTAAATCTTTGTGGGATCA\n",
      "\n",
      "negative strand\n",
      "PAM: CCC\n",
      "Protospacer: TCTTTAAGCGGCTGCCTGCGAATTACCGTC\n",
      "\n",
      "negative strand\n",
      "PAM: CCT\n",
      "Protospacer: CTTTAAGCGGCTGCCTGCGAATTACCGTCG\n",
      "\n",
      "positive strand\n",
      "PAM: CGG\n",
      "Protospacer: ACGTGATCAGAATAAACAACCCTCTTTAAG\n",
      "\n",
      "positive strand\n",
      "PAM: CGG\n",
      "Protospacer: CTCTTTAAGCGGCTGCCTGCGAATTACCGT\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Validation for His -- CRISPathBrick paper/Our paper His protospacer target = tctttaagcggctgcctgcgaattaccgtc\n",
    "targetSeq =  myDFrame2.loc[91, ['Sequence']][0]\n",
    "print \"Target Sequence: \" + targetSeq + \"\\n\"\n",
    "if(type(targetSeq) == str):\n",
    "    for base in range(len(targetSeq)):\n",
    "        if (targetSeq[base-2:base] == 'GG') & (base-(protospacer_length+PAM_length)>-1):\n",
    "            print 'positive strand'\n",
    "            print \"PAM: \" + targetSeq[base-PAM_length:base]\n",
    "            print \"Protospacer: \" + targetSeq[base-(protospacer_length+PAM_length):base-PAM_length] + \"\\n\"\n",
    "        elif (targetSeq[base-3:base-1] == 'CC') & (base+(protospacer_length)<len(targetSeq)):\n",
    "            print 'negative strand'\n",
    "            print \"PAM: \" + targetSeq[base-PAM_length:base]\n",
    "            print \"Protospacer: \" + targetSeq[base:base+(protospacer_length)] + \"\\n\"\n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of promoters annotated in E. coli: 3841\n",
      "Average number of protospacers identified per promoter (searching from 60 bps upstream and 60 bps downstream): 7\n"
     ]
    }
   ],
   "source": [
    "## Important numbers:\n",
    "\n",
    "numPromoters = myDFrame2.shape[0]\n",
    "\n",
    "print \"Number of promoters annotated in E. coli: \" + str(numPromoters)\n",
    "print \"Average number of protospacers identified per promoter (searching from \" + str(upstream) + \" bps upstream and \" + str(downstream) + \" bps downstream): \" + str(len(protospacers)/numPromoters)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda env:py27]",
   "language": "python",
   "name": "conda-env-py27-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
