{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ImportError",
     "evalue": "No module named biopython",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-c9d210ac645b>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;31m#from HTMLParser import HTMLParser\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;31m#from lxml import html\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0;32mimport\u001b[0m \u001b[0mbiopython\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      8\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      9\u001b[0m \u001b[0;31m# this function finds all pathways associated with the molecule of interest. needs work on front end to\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: No module named biopython"
     ]
    }
   ],
   "source": [
    "from IPython.display import display, Image\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "#import pythoncyc as pc\n",
    "#from HTMLParser import HTMLParser\n",
    "#from lxml import html\n",
    "#import biopython\n",
    "\n",
    "# this function finds all pathways associated with the molecule of interest. needs work on front end to \n",
    "# allow user to define G.O. function associated with molecule.\n",
    "class organismBreadBoard:\n",
    "    def __init__(self, name):\n",
    "        self.name = name\n",
    "    def pathwayFromCompound(self, compound):\n",
    "        self.compound = compound\n",
    "        pathWays = {}\n",
    "    #request pathway xml from ecocyc server\n",
    "        resp = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-compound&id=ECOLI:{}\".format(self.compound))\n",
    "    #builds element tree from the response xml text file\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #pulls all elements associated with Pathway tag\n",
    "        pathways = root.findall(\"./Pathway\")\n",
    "    #pulls all pathway IDs\n",
    "        for pathway in pathways:\n",
    "            pathID = pathway.attrib['frameid']\n",
    "            pathWays[pathID] = pathway.findall(\"common-name\")[0].text\n",
    "        #print (pathWays)\n",
    "    #return dictionary with pathway ID: common name as key:value pairs\n",
    "    #print pathWays\n",
    "        return pathWays\n",
    "\n",
    "# this function pulls all geneIDs associated with the pathwayID passed into it\n",
    "    def genesFromPathway(self, pathwayID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=genes-of-pathway&id=ECOLI:{}&detail=full\".format(pathwayID))\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        genes = root.findall(\"./Gene\")\n",
    "        genesDict = {}\n",
    "        for gene in genes:\n",
    "            geneID = gene.attrib['frameid']\n",
    "            geneName = gene.findall(\"common-name\")[0].text\n",
    "            geneStrand= gene.findall(\"transcription-direction\")[0].text\n",
    "            #print geneStrand\n",
    "        #print \"Gene: \" + geneName + \", GeneID: \" + geneID\n",
    "        #page = requests.get(\"http://biocyc.org/gene?orgid=ECOLI&id=EG11879\")\n",
    "            genesDict[geneID] = [geneName, geneStrand]\n",
    "        return genesDict\n",
    "# this function pulls the promoterIDs associated with the geneID passed into it\n",
    "    def getPromoter(self, geneID):\n",
    "        promoterDict = {}\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "        for promoter in promoters:\n",
    "            pID = promoter.attrib['frameid']\n",
    "            absPositPlusOne = promoter.find('absolute-plus-1-pos').text\n",
    "            promoterDict[pID]= absPositPlusOne\n",
    "        return promoterDict\n",
    "class FastAreader :\n",
    "    '''\n",
    "    Class to provide reading of a file containing one or more FASTA\n",
    "    formatted sequences:\n",
    "    object instantiation:\n",
    "    FastAreader(<file name>):\n",
    " \n",
    "    object attributes:\n",
    "    fname: the initial file name\n",
    " \n",
    "    methods:\n",
    "    readFasta() : returns header and sequence as strings.\n",
    "    Author: David Bernick\n",
    "    Date: April 19, 2013\n",
    "    '''\n",
    "    def __init__ (self, fname):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    " \n",
    "    def readFasta (self):\n",
    "        '''\n",
    "        using filename given in init, returns each included FastA record\n",
    "        as 2 strings - header and sequence.\n",
    "        whitespace is removed, no adjustment is made to sequence contents.\n",
    "        The initial '>' is removed from the header.\n",
    "        '''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with open(self.fname) as fileH:\n",
    "            # initialize return containers\n",
    "            header = ''\n",
    "            sequence = ''\n",
    " \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    " \n",
    "            # header is saved, get the rest of the sequence\n",
    "            # up until the next header is found\n",
    "            # then yield the results and wait for the next call.\n",
    "            # next call will resume at the yield point\n",
    "            # which is where we have the next header\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        # final header and sequence will be seen with an end of file\n",
    "        # with clause will terminate, so we do the final yield of the data\n",
    "        yield header,sequence\n",
    "        \n",
    "\n",
    "myOrg = organismBreadBoard(\"ECOLI\")\n",
    "myReader = FastAreader ('ecoli_v3_u00096_3.fa');\n",
    "\n",
    "for head, seq in myReader.readFasta() :\n",
    "    #print (head,seq)\n",
    "    targetHead=head\n",
    "    targetSeq= seq\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "allPathWays = myOrg.pathwayFromCompound(\"HIS\")\n",
    "\n",
    "for pathway in allPathWays:\n",
    "    operonDirection=\"\"\n",
    "    print allPathWays[pathway]\n",
    "    userPath = raw_input(\"Do you want to knockdown this function? yes or no.\")\n",
    "    if userPath == \"yes\":\n",
    "        genes = myOrg.genesFromPathway(pathway)\n",
    "        operonPromoterDict={}\n",
    "        for gene in genes:\n",
    "            operonDirection=genes[gene][1]\n",
    "            #print genes[gene], genes[gene][1], gene\n",
    "            #print genes[gene][0]\n",
    "            promoters_gene=myOrg.getPromoter(gene)\n",
    "            dwnStrmPromo_gene=0\n",
    "            for promoter in promoters_gene:\n",
    "                if promoters_gene[max(promoters_gene, key=promoters_gene.get)]>dwnStrmPromo_gene:\n",
    "                    dwnStrmPromo_gene = promoter\n",
    "                    dwnStrmPromoPos_gene = promoters_gene[max(promoters_gene, key=promoters_gene.get)]\n",
    "                else:\n",
    "                    continue\n",
    "            operonPromoterDict[dwnStrmPromo_gene]= dwnStrmPromoPos_gene\n",
    "        #print operonPromoterDict\n",
    "        dwnStrmPromoPos_operon=0\n",
    "        dwnStrmPromoID_operon=\"\"\n",
    "        for promoter_operon in operonPromoterDict:\n",
    "            #print operonPromoterDict[promoter_operon]\n",
    "            if int(operonPromoterDict[promoter_operon]) > dwnStrmPromoPos_operon:\n",
    "                dwnStrmPromoPos_operon = int(operonPromoterDict[promoter_operon])\n",
    "                dwnStrmPromoID_operon = promoter_operon\n",
    "            else:\n",
    "                continue\n",
    "            #print operonPromoterDict[gene][1]\n",
    "        #print dwnStrmPromoID_operon, dwnStrmPromoPos_operon\n",
    "        #print max(operonPromoterDict, key=operonPromoterDict.get)\n",
    "        break\n",
    "    elif userPath == \"no\":\n",
    "        continue\n",
    "    else:\n",
    "        print \"please enter a valid response. yes or no\"\n",
    "        continue\n",
    "if operonDirection == \"-\":\n",
    "    targetPromoSeq = targetSeq[dwnStrmPromoPos_operon:(dwnStrmPromoPos_operon+50)]\n",
    "else:\n",
    "    targetPromoSeq = targetSeq[(dwnStrmPromoPos_operon-50):dwnStrmPromoPos_operon]\n",
    "print dwnStrmPromoID_operon\n",
    "print targetPromoSeq\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HISTSYN-PWY\n",
      "Pathway:  L-histidine biosynthesis\n"
     ]
    }
   ],
   "source": [
    "from IPython.display import display, Image\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "import pythoncyc as pc\n",
    "from HTMLParser import HTMLParser\n",
    "from lxml import html\n",
    "#import biopython\n",
    "\n",
    "# this function finds all pathways associated with the molecule of interest. needs work on front end to \n",
    "# allow user to define G.O. function associated with molecule.\n",
    "def pathwayFromCompound(compound):\n",
    "    pathIDs = []\n",
    "    #request pathway xml from ecocyc server\n",
    "    resp = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-compound&id=ECOLI:{}\".format(compound))\n",
    "    #builds element tree from the response xml text file\n",
    "    root = ET.fromstring(resp.text)\n",
    "    #pulls all elements associated with Pathway tag\n",
    "    pathways = root.findall(\"./Pathway\")\n",
    "    #pulls all pathway IDs\n",
    "    for pathway in pathways:\n",
    "        pathID = pathway.attrib['frameid']\n",
    "        print (pathID)\n",
    "        print \"Pathway: \", pathway.findall(\"common-name\")[0].text\n",
    "        #instead of this we can present a tree of the biokernel\n",
    "        userPath = raw_input(\"Do you want to knockdown this function? yes or no.\")\n",
    "        if userPath == \"yes\":\n",
    "            genesFromPathway(pathID)\n",
    "            break\n",
    "        else:\n",
    "            continue\n",
    "\n",
    "# this function pulls all geneIDs associated with the pathwayID passed into it\n",
    "def genesFromPathway(pathwayID):\n",
    "    resp  = requests.get(\"http://ecocyc.org/apixml?fn=genes-of-pathway&id=ECOLI:{}\".format(pathwayID))\n",
    "    root = ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "    genes = root.findall(\"./Gene\")\n",
    "    \n",
    "    for gene in genes:\n",
    "        geneID = gene.attrib['frameid']\n",
    "        geneName = gene.findall(\"common-name\")[0].text\n",
    "        print \"Gene: \" + geneName + \", GeneID: \" + geneID\n",
    "        page = requests.get(\"http://biocyc.org/gene?orgid=ECOLI&id=EG11879\")\n",
    "        #print(page.content)\n",
    "        #tree = html.fromstring(page.content)\n",
    "        #mapPost = tree.find(\"./Map Position\")\n",
    "        #print(mapPost)\n",
    "        #root1 = ET.fromstring(resp1.text)\n",
    "        \n",
    "        getPromoter(geneID)\n",
    "\n",
    "# this function pulls the promoterIDs associated with the geneID passed into it\n",
    "def getPromoter(geneID):\n",
    "    resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}\".format(geneID))\n",
    "    root=ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "    promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "    #print promoters\n",
    "    for promoter in promoters:\n",
    "        if promoter == \"'NoneType'\":\n",
    "            print \"NoneType\"\n",
    "        else:\n",
    "            pID = promoter.attrib['frameid']\n",
    "            absPositPlusOne = promoter.find('absolute-plus-1-pos').text\n",
    "        #regulon = ecoli.containing_chromosome(pID)\n",
    "            print \"promoter: \",pID\n",
    "            print(absPositPlusOne)\n",
    "        #print(regulon)\n",
    "\n",
    "#ecoli = pc.select_organism(\"ECOLI\")\n",
    "pathwayFromCompound(\"HIS\")\n",
    "#genesFromPathway(\"HISTSYN-PWY\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from IPython.display import display, Image\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "import pythoncyc as pc\n",
    "from HTMLParser import HTMLParser\n",
    "from lxml import html\n",
    "#import biopython\n",
    "\n",
    "# this function finds all pathways associated with the molecule of interest. needs work on front end to \n",
    "# allow user to define G.O. function associated with molecule.\n",
    "def pathwayFromCompound(compound):\n",
    "    pathWays = {}\n",
    "    #request pathway xml from ecocyc server\n",
    "    resp = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-compound&id=ECOLI:{}\".format(compound))\n",
    "    #builds element tree from the response xml text file\n",
    "    root = ET.fromstring(resp.text)\n",
    "    #pulls all elements associated with Pathway tag\n",
    "    pathways = root.findall(\"./Pathway\")\n",
    "    #pulls all pathway IDs\n",
    "    for pathway in pathways:\n",
    "        pathID = pathway.attrib['frameid']\n",
    "        pathWays[pathID] = pathway.findall(\"common-name\")[0].text\n",
    "        #print (pathWays)\n",
    "    #return dictionary with pathway ID: common name as key:value pairs\n",
    "    #print pathWays\n",
    "    return pathWays\n",
    "\n",
    "# this function pulls all geneIDs associated with the pathwayID passed into it\n",
    "def genesFromPathway(pathwayID):\n",
    "    resp  = requests.get(\"http://ecocyc.org/apixml?fn=genes-of-pathway&id=ECOLI:{}&detail=full\".format(pathwayID))\n",
    "    root = ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "    genes = root.findall(\"./Gene\")\n",
    "    genesDict = {}\n",
    "    for gene in genes:\n",
    "        geneID = gene.attrib['frameid']\n",
    "        geneName = gene.findall(\"common-name\")[0].text\n",
    "        geneStrand= gene.findall(\"transcription-direction\")[0].text\n",
    "            #print geneStrand\n",
    "        #print \"Gene: \" + geneName + \", GeneID: \" + geneID\n",
    "        #page = requests.get(\"http://biocyc.org/gene?orgid=ECOLI&id=EG11879\")\n",
    "        genesDict[geneID] = [geneName, geneStrand]\n",
    "    return genesDict\n",
    "# this function pulls the promoterIDs associated with the geneID passed into it\n",
    "def getPromoter(geneID):\n",
    "    promoterDict = {}\n",
    "    resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "    root=ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "    promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "    #transcripts=root.findall(\"./Transcription-Unit\")\n",
    "    #for TU in transcripts:\n",
    "        #TUID=TU.attrib['frameid']\n",
    "        #print TUID\n",
    "        #resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-unit-genes&id=ECOLI:{}&detail=full\".format(TUID))\n",
    "        #root = ET.fromstring(resp.text)\n",
    "        #print resp.text\n",
    "        #resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-unit-promoter&id=ECOLI:{}\".format(TUID))\n",
    "        #root = ET.fromstring(resp.text)\n",
    "        #print resp.text\n",
    "    #print promoters\n",
    "    for promoter in promoters:\n",
    "        pID = promoter.attrib['frameid']\n",
    "        absPositPlusOne = promoter.find('absolute-plus-1-pos').text\n",
    "        #regulon = ecoli.containing_chromosome(pID)\n",
    "        #print \"promoter: \",pID\n",
    "        promoterDict[pID]= absPositPlusOne\n",
    "        #print(regulon)\n",
    "    return promoterDict\n",
    "\n",
    "        \n",
    "#EG10444\n",
    "#TU0-6650\n",
    "#ecoli = pc.select_organism(\"ECOLI\")\n",
    "allPathways = pathwayFromCompound(\"HIS\")\n",
    "#genesFromPathway(\"HISTSYN-PWY\")\n",
    "\n",
    "#getPromoter('EG10444')\n",
    "\n",
    "for pathway in allPathways:\n",
    "    print allPathways[pathway]\n",
    "    userPath = raw_input(\"Do you want to knockdown this function? yes or no.\")\n",
    "    if userPath == \"yes\":\n",
    "        genes = genesFromPathway(pathway)\n",
    "        for gene in genes:\n",
    "            print genes[gene], gene\n",
    "            promoters=getPromoter(gene)\n",
    "        #print promoters\n",
    "        break\n",
    "    else:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<type 'dict'>\n",
      "tRNA charging\n",
      "Do you want to knockdown this function? yes or no.no\n",
      "superpathway of histidine, purine, and pyrimidine biosynthesis\n",
      "Do you want to knockdown this function? yes or no.no\n",
      "L-histidine biosynthesis\n",
      "Do you want to knockdown this function? yes or no.yes\n",
      "['hisI', '+'] + EG10451\n",
      "['hisH', '+'] + EG10450\n",
      "['hisC', '+'] + EG10446\n",
      "['hisD', '+'] + EG10447\n",
      "['hisA', '+'] + EG10444\n",
      "['hisB', '+'] + EG10445\n",
      "['hisF', '+'] + EG10448\n",
      "['hisG', '+'] + EG10449\n",
      "PM00398 2093360\n"
     ]
    }
   ],
   "source": [
    "from IPython.display import display, Image\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "import pythoncyc as pc\n",
    "from HTMLParser import HTMLParser\n",
    "from lxml import html\n",
    "#import biopython\n",
    "\n",
    "# this function finds all pathways associated with the molecule of interest. needs work on front end to \n",
    "# allow user to define G.O. function associated with molecule.\n",
    "class organismBreadBoard:\n",
    "    def __init__(self, name):\n",
    "        self.name = name\n",
    "    def pathwayFromCompound(self, compound):\n",
    "        self.compound = compound\n",
    "        pathWays = {}\n",
    "    #request pathway xml from ecocyc server\n",
    "        resp = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-compound&id=ECOLI:{}\".format(self.compound))\n",
    "    #builds element tree from the response xml text file\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #pulls all elements associated with Pathway tag\n",
    "        pathways = root.findall(\"./Pathway\")\n",
    "    #pulls all pathway IDs\n",
    "        for pathway in pathways:\n",
    "            pathID = pathway.attrib['frameid']\n",
    "            pathWays[pathID] = pathway.findall(\"common-name\")[0].text\n",
    "        #print (pathWays)\n",
    "    #return dictionary with pathway ID: common name as key:value pairs\n",
    "    #print pathWays\n",
    "        return pathWays\n",
    "\n",
    "# this function pulls all geneIDs associated with the pathwayID passed into it\n",
    "    def genesFromPathway(self, pathwayID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=genes-of-pathway&id=ECOLI:{}&detail=full\".format(pathwayID))\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        genes = root.findall(\"./Gene\")\n",
    "        genesDict = {}\n",
    "        for gene in genes:\n",
    "            geneID = gene.attrib['frameid']\n",
    "            geneName = gene.findall(\"common-name\")[0].text\n",
    "            geneStrand= gene.findall(\"transcription-direction\")[0].text\n",
    "            #print geneStrand\n",
    "        #print \"Gene: \" + geneName + \", GeneID: \" + geneID\n",
    "        #page = requests.get(\"http://biocyc.org/gene?orgid=ECOLI&id=EG11879\")\n",
    "            genesDict[geneID] = [geneName, geneStrand]\n",
    "        return genesDict\n",
    "# this function pulls the promoterIDs associated with the geneID passed into it\n",
    "    def getPromoter(self, geneID):\n",
    "        promoterDict = {}\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "        for promoter in promoters:\n",
    "            pID = promoter.attrib['frameid']\n",
    "            absPositPlusOne = promoter.find('absolute-plus-1-pos').text\n",
    "            promoterDict[pID]= absPositPlusOne\n",
    "        return promoterDict\n",
    "\n",
    "        \n",
    "#EG10444\n",
    "#TU0-6650\n",
    "#ecoli = pc.select_organism(\"ECOLI\")\n",
    "#allPathWays=pathwayFromCompound(\"HIS\").pathwayFromCompound()\n",
    "myOrg = organismBreadBoard(\"ECOLI\")\n",
    "allPathWays = myOrg.pathwayFromCompound(\"HIS\")\n",
    "print type(allPathWays)\n",
    "#genesFromPathway(\"HISTSYN-PWY\")\n",
    "\n",
    "#getPromoter('EG10444')\n",
    "\n",
    "for pathway in allPathWays:\n",
    "    print allPathWays[pathway]\n",
    "    userPath = raw_input(\"Do you want to knockdown this function? yes or no.\")\n",
    "    if userPath == \"yes\":\n",
    "        genes = myOrg.genesFromPathway(pathway)\n",
    "        operonPromoterDict={}\n",
    "        for gene in genes:\n",
    "            print genes[gene], genes[gene][1], gene\n",
    "            #print genes[gene][0]\n",
    "            promoters_gene=myOrg.getPromoter(gene)\n",
    "            dwnStrmPromo_gene=0\n",
    "            for promoter in promoters_gene:\n",
    "                if promoters_gene[max(promoters_gene, key=promoters_gene.get)]>dwnStrmPromo_gene:\n",
    "                    dwnStrmPromo_gene = promoter\n",
    "                    dwnStrmPromoPos_gene = promoters_gene[max(promoters_gene, key=promoters_gene.get)]\n",
    "                else:\n",
    "                    continue\n",
    "            operonPromoterDict[dwnStrmPromo_gene]= dwnStrmPromoPos_gene\n",
    "        #print operonPromoterDict\n",
    "        dwnStrmPromoPos_operon=0\n",
    "        dwnStrmPromoID_operon=\"\"\n",
    "        for promoter_operon in operonPromoterDict:\n",
    "            if operonPromoterDict[promoter_operon] > dwnStrmPromoPos_operon:\n",
    "                dwnStrmPromoPos_operon = operonPromoterDict[promoter_operon]\n",
    "                dwnStrmPromoID_operon = promoter_operon\n",
    "            else:\n",
    "                continue\n",
    "            #print operonPromoterDict[gene][1]\n",
    "        print dwnStrmPromoID_operon, dwnStrmPromoPos_operon\n",
    "        #print max(operonPromoterDict, key=operonPromoterDict.get)\n",
    "        break\n",
    "    else:\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'myReader' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-e6ec950cc8bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[0;31m# presumed object instantiation and example usage\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[0;31m#myReader = FastAreader ('ecoli_v3_u00096_3.fa');\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 59\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mhead\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mseq\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mmyReader\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadFasta\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     60\u001b[0m     \u001b[0;32mprint\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'myReader' is not defined"
     ]
    }
   ],
   "source": [
    "class FastAreader :\n",
    "    '''\n",
    "    Class to provide reading of a file containing one or more FASTA\n",
    "    formatted sequences:\n",
    "    object instantiation:\n",
    "    FastAreader(<file name>):\n",
    " \n",
    "    object attributes:\n",
    "    fname: the initial file name\n",
    " \n",
    "    methods:\n",
    "    readFasta() : returns header and sequence as strings.\n",
    "    Author: David Bernick\n",
    "    Date: April 19, 2013\n",
    "    '''\n",
    "    def __init__ (self, fname):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    " \n",
    "    def readFasta (self):\n",
    "        '''\n",
    "        using filename given in init, returns each included FastA record\n",
    "        as 2 strings - header and sequence.\n",
    "        whitespace is removed, no adjustment is made to sequence contents.\n",
    "        The initial '>' is removed from the header.\n",
    "        '''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with open(self.fname) as fileH:\n",
    "            # initialize return containers\n",
    "            header = ''\n",
    "            sequence = ''\n",
    " \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    " \n",
    "            # header is saved, get the rest of the sequence\n",
    "            # up until the next header is found\n",
    "            # then yield the results and wait for the next call.\n",
    "            # next call will resume at the yield point\n",
    "            # which is where we have the next header\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        # final header and sequence will be seen with an end of file\n",
    "        # with clause will terminate, so we do the final yield of the data\n",
    "        yield header,sequence\n",
    " \n",
    "# presumed object instantiation and example usage\n",
    "#myReader = FastAreader ('ecoli_v3_u00096_3.fa');\n",
    "for head, seq in myReader.readFasta() :\n",
    "    print (head,seq)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#for python 3.5\n",
    "from IPython.display import display, Image\n",
    "import xml.etree.ElementTree as ET\n",
    "import requests\n",
    "#import pythoncyc as pc\n",
    "#from HTMLParser import HTMLParser\n",
    "#from lxml import html\n",
    "import biopython\n",
    "\n",
    "# this function finds all pathways associated with the molecule of interest. needs work on front end to \n",
    "# allow user to define G.O. function associated with molecule.\n",
    "class organismBreadBoard:\n",
    "    def __init__(self, name):\n",
    "        self.name = name\n",
    "    def pathwayFromCompound(self, compound):\n",
    "        self.compound = compound\n",
    "        pathWays = {}\n",
    "    #request pathway xml from ecocyc server\n",
    "        resp = requests.get(\"http://ecocyc.org/apixml?fn=pathways-of-compound&id=ECOLI:{}\".format(self.compound))\n",
    "    #builds element tree from the response xml text file\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #pulls all elements associated with Pathway tag\n",
    "        pathways = root.findall(\"./Pathway\")\n",
    "    #pulls all pathway IDs\n",
    "        for pathway in pathways:\n",
    "            pathID = pathway.attrib['frameid']\n",
    "            pathWays[pathID] = pathway.findall(\"common-name\")[0].text\n",
    "        #print (pathWays)\n",
    "    #return dictionary with pathway ID: common name as key:value pairs\n",
    "    #print pathWays\n",
    "        return pathWays\n",
    "\n",
    "# this function pulls all geneIDs associated with the pathwayID passed into it\n",
    "    def genesFromPathway(self, pathwayID):\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=genes-of-pathway&id=ECOLI:{}&detail=full\".format(pathwayID))\n",
    "        root = ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        genes = root.findall(\"./Gene\")\n",
    "        genesDict = {}\n",
    "        for gene in genes:\n",
    "            geneID = gene.attrib['frameid']\n",
    "            geneName = gene.findall(\"common-name\")[0].text\n",
    "            geneStrand= gene.findall(\"transcription-direction\")[0].text\n",
    "            #print geneStrand\n",
    "        #print \"Gene: \" + geneName + \", GeneID: \" + geneID\n",
    "        #page = requests.get(\"http://biocyc.org/gene?orgid=ECOLI&id=EG11879\")\n",
    "            genesDict[geneID] = [geneName, geneStrand]\n",
    "        return genesDict\n",
    "# this function pulls the promoterIDs associated with the geneID passed into it\n",
    "    def getPromoter(self, geneID):\n",
    "        promoterDict = {}\n",
    "        resp  = requests.get(\"http://ecocyc.org/apixml?fn=transcription-units-of-gene&id=ECOLI:{}&detail=full\".format(geneID))\n",
    "        root=ET.fromstring(resp.text)\n",
    "    #print(resp.text)\n",
    "        promoters=root.findall(\"./Transcription-Unit/component/Promoter\")\n",
    "        for promoter in promoters:\n",
    "            pID = promoter.attrib['frameid']\n",
    "            absPositPlusOne = promoter.find('absolute-plus-1-pos').text\n",
    "            promoterDict[pID]= absPositPlusOne\n",
    "        return promoterDict\n",
    "class FastAreader :\n",
    "    '''\n",
    "    Class to provide reading of a file containing one or more FASTA\n",
    "    formatted sequences:\n",
    "    object instantiation:\n",
    "    FastAreader(<file name>):\n",
    " \n",
    "    object attributes:\n",
    "    fname: the initial file name\n",
    " \n",
    "    methods:\n",
    "    readFasta() : returns header and sequence as strings.\n",
    "    Author: David Bernick\n",
    "    Date: April 19, 2013\n",
    "    '''\n",
    "    def __init__ (self, fname):\n",
    "        '''contructor: saves attribute fname '''\n",
    "        self.fname = fname\n",
    " \n",
    "    def readFasta (self):\n",
    "        '''\n",
    "        using filename given in init, returns each included FastA record\n",
    "        as 2 strings - header and sequence.\n",
    "        whitespace is removed, no adjustment is made to sequence contents.\n",
    "        The initial '>' is removed from the header.\n",
    "        '''\n",
    "        header = ''\n",
    "        sequence = ''\n",
    "        \n",
    "        with open(self.fname) as fileH:\n",
    "            # initialize return containers\n",
    "            header = ''\n",
    "            sequence = ''\n",
    " \n",
    "            # skip to first fasta header\n",
    "            line = fileH.readline()\n",
    "            while not line.startswith('>') :\n",
    "                line = fileH.readline()\n",
    "            header = line[1:].rstrip()\n",
    " \n",
    "            # header is saved, get the rest of the sequence\n",
    "            # up until the next header is found\n",
    "            # then yield the results and wait for the next call.\n",
    "            # next call will resume at the yield point\n",
    "            # which is where we have the next header\n",
    "            for line in fileH:\n",
    "                if line.startswith ('>'):\n",
    "                    yield header,sequence\n",
    "                    header = line[1:].rstrip()\n",
    "                    sequence = ''\n",
    "                else :\n",
    "                    sequence += ''.join(line.rstrip().split()).upper()\n",
    "        # final header and sequence will be seen with an end of file\n",
    "        # with clause will terminate, so we do the final yield of the data\n",
    "        yield header,sequence\n",
    "        \n",
    "\n",
    "myOrg = organismBreadBoard(\"ECOLI\")\n",
    "myReader = FastAreader ('ecoli_v3_u00096_3.fa');\n",
    "\n",
    "for head, seq in myReader.readFasta() :\n",
    "    #print (head,seq)\n",
    "    targetHead=head\n",
    "    targetSeq= seq\n",
    "\n",
    "\n",
    "\n",
    "    \n",
    "allPathWays = myOrg.pathwayFromCompound(\"HIS\")\n",
    "\n",
    "for pathway in allPathWays:\n",
    "    operonDirection=\"\"\n",
    "    print (allPathWays[pathway])\n",
    "    userPath = raw_input(\"Do you want to knockdown this function? yes or no.\")\n",
    "    if userPath == \"yes\":\n",
    "        genes = myOrg.genesFromPathway(pathway)\n",
    "        operonPromoterDict={}\n",
    "        for gene in genes:\n",
    "            operonDirection=genes[gene][1]\n",
    "            #print genes[gene], genes[gene][1], gene\n",
    "            #print genes[gene][0]\n",
    "            promoters_gene=myOrg.getPromoter(gene)\n",
    "            dwnStrmPromo_gene=0\n",
    "            for promoter in promoters_gene:\n",
    "                if promoters_gene[max(promoters_gene, key=promoters_gene.get)]>dwnStrmPromo_gene:\n",
    "                    dwnStrmPromo_gene = promoter\n",
    "                    dwnStrmPromoPos_gene = promoters_gene[max(promoters_gene, key=promoters_gene.get)]\n",
    "                else:\n",
    "                    continue\n",
    "            operonPromoterDict[dwnStrmPromo_gene]= dwnStrmPromoPos_gene\n",
    "        #print operonPromoterDict\n",
    "        dwnStrmPromoPos_operon=0\n",
    "        dwnStrmPromoID_operon=\"\"\n",
    "        for promoter_operon in operonPromoterDict:\n",
    "            #print operonPromoterDict[promoter_operon]\n",
    "            if int(operonPromoterDict[promoter_operon]) > dwnStrmPromoPos_operon:\n",
    "                dwnStrmPromoPos_operon = int(operonPromoterDict[promoter_operon])\n",
    "                dwnStrmPromoID_operon = promoter_operon\n",
    "            else:\n",
    "                continue\n",
    "            #print operonPromoterDict[gene][1]\n",
    "        #print dwnStrmPromoID_operon, dwnStrmPromoPos_operon\n",
    "        #print max(operonPromoterDict, key=operonPromoterDict.get)\n",
    "        break\n",
    "    elif userPath == \"no\":\n",
    "        continue\n",
    "    else:\n",
    "        print (\"please enter a valid response. yes or no\")\n",
    "        continue\n",
    "if operonDirection == \"-\":\n",
    "    targetPromoSeq = targetSeq[(dwnStrmPromoPos_operon-50):dwnStrmPromoPos_operon]\n",
    "else:\n",
    "    targetPromoSeq = targetSeq[(dwnStrmPromoPos_operon-50):dwnStrmPromoPos_operon]\n",
    "print (dwnStrmPromoID_operon, operonDirection)\n",
    "print (targetPromoSeq)\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [py27]",
   "language": "python",
   "name": "Python [py27]"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
